{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fe28a16c-75e1-448c-939c-b1b221f62321",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data analysis packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# file and command line packages\n",
    "import os\n",
    "import subprocess\n",
    "import gzip\n",
    "\n",
    "# string parsing\n",
    "import re\n",
    "\n",
    "# sparse matrix packages\n",
    "from scipy.io import mmread\n",
    "\n",
    "# modeling\n",
    "from scipy.stats import nbinom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a93eaa8-6669-4cf2-bc81-300df8ad9fd5",
   "metadata": {},
   "source": [
    "### Load in Group and Sequence Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2befb927-2093-49ea-923a-8a8cc82d62c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF593_TSS</td>\n",
       "      <td>AACAGCCCGGCCGGCCAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATPIF1_TSS</td>\n",
       "      <td>AACGAGAGACTGCTTGCTGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG\n",
       "1  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG\n",
       "2    FAM83A_TSS  AACACACCACGGAGGAGTGG\n",
       "3    ZNF593_TSS  AACAGCCCGGCCGGCCAAGG\n",
       "4    ATPIF1_TSS  AACGAGAGACTGCTTGCTGG"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load in group sequence dataframe\n",
    "group_sequence_df = pd.read_csv('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_grna_groups.at_scale.txt', \n",
    "                                sep='\\t', names=['gRNA_group', 'spacer_sequence'])\n",
    "group_sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4562297d-83b5-47dd-a90a-c8e046b625bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13189, 2)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check dataframe shape\n",
    "group_sequence_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dcd19eda-6258-4e92-9091-792da47ff059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gRNA_group         0\n",
       "spacer_sequence    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check dataframe missingness\n",
    "group_sequence_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b6ed2a53-4a3d-4b8f-823b-9d9542cec18e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13189, 2)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check to ensure that there are no duplicates in the file\n",
    "group_sequence_df.drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30363d39-cc85-4f9f-9d9b-04009253400f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if there are two guides per gRNA group\n",
    "(group_sequence_df.groupby('gRNA_group').count()['spacer_sequence'] != 2).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "402de820-146a-4d78-b436-68e4b2d39cf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gRNA_group</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bassik_mch</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_Klannchr1_HBG1_HBG1_tss_both</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_Klannchr1_HS3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_Klannchr1_HS4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_Klannchr_HS1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_mosaic_HB_HBE1_tss_A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_control_mosaic_HB_HBE1_tss_B</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          spacer_sequence\n",
       "gRNA_group                                               \n",
       "bassik_mch                                              1\n",
       "pos_control_Klannchr1_HBG1_HBG1_tss_both                1\n",
       "pos_control_Klannchr1_HS3                               1\n",
       "pos_control_Klannchr1_HS4                               1\n",
       "pos_control_Klannchr_HS1                                1\n",
       "pos_control_mosaic_HB_HBE1_tss_A                        1\n",
       "pos_control_mosaic_HB_HBE1_tss_B                        1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# see groups where there are not 2 guides\n",
    "df = group_sequence_df.groupby('gRNA_group').count()\n",
    "df[df['spacer_sequence'] != 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcf9b520-9e91-4bdf-9fe1-8912cd77dc7d",
   "metadata": {},
   "source": [
    "### Use BLAST to determine Genomic Coordinates from Sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7592567b-b416-427f-981e-fbf819893942",
   "metadata": {},
   "source": [
    "Using BLAST to determine the genomic coordinates involves some preparation steps. First, the reference genome assembly (GRCh37) must be converted from a FASTA file to a BLAST database for alignment. This is because this genome assembly is not provided in the preformatted BLAST databases. This can be done using the makeblastdb command. After this db is created, it can now be used to determine the genomic coordinates.\n",
    "\n",
    "There are many parameters that are specified in the BLAST search. The db parameter is a path to where the db is stored (where makeblastdb was run) plus \"/hg19\". The query parameter is the input FASTA file, which contains the sequence to be queried by the BLAST db. The task parameter is especially important, since the guide sequences are shorter, this parameter adjusts other search parameters to yield results for the guide search (without this parameter, BLAST will not find any hits!). The outfmt parameter tells BLAST to return tabular output (specified by 6), and states the specific columns that BLAST should return from the search. This is for easier parsing using pandas. Finally, the out parameter is the filepath to where the output should be stored from BLAST."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5507513b-a79f-4530-afad-637627d5c7b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\n",
      "  blastn [-h] [-help] [-import_search_strategy filename]\n",
      "    [-export_search_strategy filename] [-task task_name] [-db database_name]\n",
      "    [-dbsize num_letters] [-gilist filename] [-seqidlist filename]\n",
      "    [-negative_gilist filename] [-entrez_query entrez_query]\n",
      "    [-db_soft_mask filtering_algorithm] [-db_hard_mask filtering_algorithm]\n",
      "    [-subject subject_input_file] [-subject_loc range] [-query input_file]\n",
      "    [-out output_file] [-evalue evalue] [-word_size int_value]\n",
      "    [-gapopen open_penalty] [-gapextend extend_penalty]\n",
      "    [-perc_identity float_value] [-qcov_hsp_perc float_value]\n",
      "    [-max_hsps int_value] [-xdrop_ungap float_value] [-xdrop_gap float_value]\n",
      "    [-xdrop_gap_final float_value] [-searchsp int_value]\n",
      "    [-sum_stats bool_value] [-penalty penalty] [-reward reward] [-no_greedy]\n",
      "    [-min_raw_gapped_score int_value] [-template_type type]\n",
      "    [-template_length int_value] [-dust DUST_options]\n",
      "    [-filtering_db filtering_database]\n",
      "    [-window_masker_taxid window_masker_taxid]\n",
      "    [-window_masker_db window_masker_db] [-soft_masking soft_masking]\n",
      "    [-ungapped] [-culling_limit int_value] [-best_hit_overhang float_value]\n",
      "    [-best_hit_score_edge float_value] [-window_size int_value]\n",
      "    [-off_diagonal_range int_value] [-use_index boolean] [-index_name string]\n",
      "    [-lcase_masking] [-query_loc range] [-strand strand] [-parse_deflines]\n",
      "    [-outfmt format] [-show_gis] [-num_descriptions int_value]\n",
      "    [-num_alignments int_value] [-line_length line_length] [-html]\n",
      "    [-max_target_seqs num_sequences] [-num_threads int_value] [-remote]\n",
      "    [-version]\n",
      "\n",
      "DESCRIPTION\n",
      "   Nucleotide-Nucleotide BLAST 2.5.0+\n",
      "\n",
      "Use '-help' to print detailed descriptions of command line arguments\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# print BLAST help menu (for reference)\n",
    "print((subprocess.run([\"blastn\", \"-h\"], capture_output=True).stdout).decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e258c791-edae-463e-9c30-fa8e519a4c7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_fasta(sequence):\n",
    "    '''helper function to create FASTA file given input sequence'''\n",
    "    \n",
    "    # open file and write sequence to BLAST folder\n",
    "    filename = sequence + '.fasta'\n",
    "    filepath = './../blast/input/' + filename\n",
    "    with open(filepath, 'w') as f:\n",
    "        f.write(sequence + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "67abdf91-5d09-4ffe-b467-2e5b805120d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_blast(file):\n",
    "    '''helper function to get genomic coordinates from BLAST output'''\n",
    "    \n",
    "    # define BLAST command line arguments\n",
    "    sequence = file.split('.')[0]\n",
    "    blast_args = [\"blastn\", \"-db\", \"../../blast/hg19\", \"-query\", '../blast/input/' + file, \"-task\", \"blastn-short\", \n",
    "                  '-outfmt', '6 sseqid evalue sstart send', '-out', '../blast/output/' + sequence + '.tsv']\n",
    "    \n",
    "    # run BLAST\n",
    "    subprocess.run(blast_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5fefdec5-92f7-4a23-8a75-1c7d18c5f2c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        None\n",
       "1        None\n",
       "2        None\n",
       "3        None\n",
       "4        None\n",
       "         ... \n",
       "13184    None\n",
       "13185    None\n",
       "13186    None\n",
       "13187    None\n",
       "13188    None\n",
       "Name: spacer_sequence, Length: 13189, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create FASTA files for each spacer sequence\n",
    "group_sequence_df['spacer_sequence'].apply(create_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f780f60b-e698-4e1a-80ee-f67535e61ff4",
   "metadata": {},
   "source": [
    "After generating the FASTA files, the next step is to run BLAST on each FASTA file to determine the genomic coordinates of the guide. BLAST can be run using the run_blast.py script under the \"scripts\" folder. This is done using the run_blast function defined above, but in a for loop for each FASTA file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "652e3959-f653-427b-92d3-9550e23682a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52946190\n",
      "52946209\n"
     ]
    }
   ],
   "source": [
    "# test getting start and end coordinates for single BLAST query\n",
    "colnames = ['chrom', 'evalue', 'start', 'end']\n",
    "hits_df = pd.read_csv('../blast/output/AAAAAAAAGGTTAAGAGTAG.tsv', sep='\\t', names=colnames)\n",
    "sorted_hits_df = hits_df.sort_values(by='evalue', ascending=True)\n",
    "print(sorted_hits_df.iloc[0]['start'])\n",
    "print(sorted_hits_df.iloc[0]['end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a5997a22-d60d-4f6f-974c-9621087f7600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF593_TSS</td>\n",
       "      <td>AACAGCCCGGCCGGCCAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATPIF1_TSS</td>\n",
       "      <td>AACGAGAGACTGCTTGCTGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG\n",
       "1  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG\n",
       "2    FAM83A_TSS  AACACACCACGGAGGAGTGG\n",
       "3    ZNF593_TSS  AACAGCCCGGCCGGCCAAGG\n",
       "4    ATPIF1_TSS  AACGAGAGACTGCTTGCTGG"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview group sequence dataframe\n",
    "group_sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "481fca17-fe90-47d7-8aef-23ef72f0d952",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomic_coordinates(seq):\n",
    "    '''helper function to get genomic coordinates from BLAST output'''\n",
    "    \n",
    "    # load in BLAST hits and sort by evalue\n",
    "    colnames = ['chrom' ,'evalue', 'start', 'end']\n",
    "    hits_df = pd.read_csv('../blast/output/' + seq + '.tsv', sep='\\t', names=colnames)\n",
    "    sorted_hits_df = hits_df.sort_values(by='evalue', ascending=True)\n",
    "    \n",
    "    # return start and end coordinates from BLAST\n",
    "    if sorted_hits_df.shape[0] == 0:\n",
    "        return [np.nan, np.nan, np.nan, np.nan]\n",
    "    else:    \n",
    "        start_coord = sorted_hits_df.iloc[0]['start']\n",
    "        end_coord = sorted_hits_df.iloc[0]['end']\n",
    "        chrom = sorted_hits_df.iloc[0]['chrom']\n",
    "        evalue = sorted_hits_df.iloc[0]['evalue']\n",
    "        return [start_coord, end_coord, chrom, evalue]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f8bdbf2f-509d-44f9-99f4-7d87970a41ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get genomic coordinates for each spacer sequence\n",
    "group_sequence_df['genomic_coords'] = group_sequence_df['spacer_sequence'].apply(get_genomic_coordinates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "700ca2f5-815c-4bea-9222-705c6e6f64c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>[26606551, 26606569, chr1, 0.032]</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>[10530445, 10530427, chr11, 0.032]</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>[124195027, 124195008, chr8, 0.008]</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF593_TSS</td>\n",
       "      <td>AACAGCCCGGCCGGCCAAGG</td>\n",
       "      <td>[26496467, 26496448, chr1, 0.008]</td>\n",
       "      <td>26496467.0</td>\n",
       "      <td>26496448.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATPIF1_TSS</td>\n",
       "      <td>AACGAGAGACTGCTTGCTGG</td>\n",
       "      <td>[28562642, 28562660, chr1, 0.032]</td>\n",
       "      <td>28562642.0</td>\n",
       "      <td>28562660.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence                       genomic_coords  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG    [26606551, 26606569, chr1, 0.032]   \n",
       "1  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   [10530445, 10530427, chr11, 0.032]   \n",
       "2    FAM83A_TSS  AACACACCACGGAGGAGTGG  [124195027, 124195008, chr8, 0.008]   \n",
       "3    ZNF593_TSS  AACAGCCCGGCCGGCCAAGG    [26496467, 26496448, chr1, 0.008]   \n",
       "4    ATPIF1_TSS  AACGAGAGACTGCTTGCTGG    [28562642, 28562660, chr1, 0.032]   \n",
       "\n",
       "         start          end  chrom  evalue  \n",
       "0   26606551.0   26606569.0   chr1   0.032  \n",
       "1   10530445.0   10530427.0  chr11   0.032  \n",
       "2  124195027.0  124195008.0   chr8   0.008  \n",
       "3   26496467.0   26496448.0   chr1   0.008  \n",
       "4   28562642.0   28562660.0   chr1   0.032  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# divide output into start and end columns\n",
    "group_sequence_df['start'] = group_sequence_df['genomic_coords'].apply(lambda x: x[0])\n",
    "group_sequence_df['end'] = group_sequence_df['genomic_coords'].apply(lambda x: x[1])\n",
    "group_sequence_df['chrom'] = group_sequence_df['genomic_coords'].apply(lambda x: x[2])\n",
    "group_sequence_df['evalue'] = group_sequence_df['genomic_coords'].apply(lambda x: x[3])\n",
    "group_sequence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "94e81f0b-01ab-471e-89b2-f5d3da13e24c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gRNA_group          0\n",
       "spacer_sequence     0\n",
       "genomic_coords      0\n",
       "start              12\n",
       "end                12\n",
       "chrom              12\n",
       "evalue             12\n",
       "dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# analyze missingness in coordinates data\n",
    "group_sequence_df.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8752093-3a86-4f45-adf3-67af1c2d22a0",
   "metadata": {},
   "source": [
    "### Validate BLAST Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2c340764-5a19-4748-bb67-bd9b4070dc4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF593_TSS</td>\n",
       "      <td>AACAGCCCGGCCGGCCAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATPIF1_TSS</td>\n",
       "      <td>AACGAGAGACTGCTTGCTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TIPRL_TSS</td>\n",
       "      <td>AACGGCTCGGAAGCCTAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MYL4_TSS</td>\n",
       "      <td>AAGAAGAAACCGAGAGACGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CNBP_TSS</td>\n",
       "      <td>AAGACGGCTCGCAAGGTAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RPS3_TSS</td>\n",
       "      <td>AAGAGGAAGGTGAGCCTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RPL23_TSS</td>\n",
       "      <td>AAGATGTCGAAGCGAGGTGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG\n",
       "1  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG\n",
       "2    FAM83A_TSS  AACACACCACGGAGGAGTGG\n",
       "3    ZNF593_TSS  AACAGCCCGGCCGGCCAAGG\n",
       "4    ATPIF1_TSS  AACGAGAGACTGCTTGCTGG\n",
       "5     TIPRL_TSS  AACGGCTCGGAAGCCTAGGG\n",
       "6      MYL4_TSS  AAGAAGAAACCGAGAGACGG\n",
       "7      CNBP_TSS  AAGACGGCTCGCAAGGTAGG\n",
       "8      RPS3_TSS  AAGAGGAAGGTGAGCCTCTG\n",
       "9     RPL23_TSS  AAGATGTCGAAGCGAGGTGG"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print first 10 sequences to validate BLAST results\n",
    "group_sequence_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b8f1c24-fa36-41af-bbac-5dfa5ddfbdca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "      <th>general_group</th>\n",
       "      <th>chr.targetgene</th>\n",
       "      <th>start.targetgene</th>\n",
       "      <th>stop.targetgene</th>\n",
       "      <th>ENSG.targetgene</th>\n",
       "      <th>targetgene_short_name</th>\n",
       "      <th>strand.targetgene</th>\n",
       "      <th>pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28034777</td>\n",
       "      <td>28034778</td>\n",
       "      <td>ENSG00000150051</td>\n",
       "      <td>MKX</td>\n",
       "      <td>-</td>\n",
       "      <td>MKX:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28287976</td>\n",
       "      <td>28287977</td>\n",
       "      <td>ENSG00000169126</td>\n",
       "      <td>ARMC4</td>\n",
       "      <td>-</td>\n",
       "      <td>ARMC4:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28571017</td>\n",
       "      <td>28571018</td>\n",
       "      <td>ENSG00000150054</td>\n",
       "      <td>MPP7</td>\n",
       "      <td>-</td>\n",
       "      <td>MPP7:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28821422</td>\n",
       "      <td>28821423</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "      <td>WAC</td>\n",
       "      <td>+</td>\n",
       "      <td>WAC:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28966271</td>\n",
       "      <td>28966272</td>\n",
       "      <td>ENSG00000095739</td>\n",
       "      <td>BAMBI</td>\n",
       "      <td>+</td>\n",
       "      <td>BAMBI:bassik_mch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gRNAgroup.chr gRNAgroup.start gRNAgroup.stop   gRNAgroup general_group  \\\n",
       "0           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "1           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "2           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "3           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "4           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "\n",
       "  chr.targetgene  start.targetgene  stop.targetgene  ENSG.targetgene  \\\n",
       "0          chr10          28034777         28034778  ENSG00000150051   \n",
       "1          chr10          28287976         28287977  ENSG00000169126   \n",
       "2          chr10          28571017         28571018  ENSG00000150054   \n",
       "3          chr10          28821422         28821423  ENSG00000095787   \n",
       "4          chr10          28966271         28966272  ENSG00000095739   \n",
       "\n",
       "  targetgene_short_name strand.targetgene             pairs  \n",
       "0                   MKX                 -    MKX:bassik_mch  \n",
       "1                 ARMC4                 -  ARMC4:bassik_mch  \n",
       "2                  MPP7                 -   MPP7:bassik_mch  \n",
       "3                   WAC                 +    WAC:bassik_mch  \n",
       "4                 BAMBI                 +  BAMBI:bassik_mch  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in dataframe of intended target regions\n",
    "dtypes = {\n",
    "    'gRNAgroup.start': str,\n",
    "    'gRNAgroup.stop': str\n",
    "}\n",
    "\n",
    "intended_targets_df = pd.read_csv('/iblm/netapp/data1/external/Gasperini2019/suppl/' + \\\n",
    "                                  'GSE120861_gene_gRNAgroup_pair_table.at_scale.txt', sep='\\t', dtype=dtypes)\n",
    "intended_targets_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6fa226bf-dcc9-48f0-8130-75fe90541bdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gRNAgroup.chr gRNAgroup.start gRNAgroup.stop   gRNAgroup\n",
       "0           NTC             NTC            NTC  bassik_mch\n",
       "1           NTC             NTC            NTC  bassik_mch\n",
       "2           NTC             NTC            NTC  bassik_mch\n",
       "3           NTC             NTC            NTC  bassik_mch\n",
       "4           NTC             NTC            NTC  bassik_mch"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select only necessary columns from the data frame\n",
    "intended_targets_df = intended_targets_df.iloc[:, 0:4]\n",
    "intended_targets_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d5d778e4-901a-4a30-a878-b9f85daaa72b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1086564, 4)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get current shape of intended targets\n",
    "intended_targets_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8abbc41e-2ad1-4eb8-a4e7-2820fe13b4fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicate values prior to merging\n",
    "intended_targets_df = intended_targets_df.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1740674e-2271-4eea-9d93-d581247b1c63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6585, 4)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get new shape of intended targets after dropping duplicates\n",
    "intended_targets_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "855209ba-af37-4dd6-8046-2787f200a52b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1086199</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5291385</td>\n",
       "      <td>5291386</td>\n",
       "      <td>pos_control_HS2_Klann_mosaic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086272</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5291385</td>\n",
       "      <td>5291386</td>\n",
       "      <td>pos_control_Klannchr1_HBG1_HBG1_tss_both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086345</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5291385</td>\n",
       "      <td>5291386</td>\n",
       "      <td>pos_control_mosaic_HB_HBE1_tss_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086418</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5291385</td>\n",
       "      <td>5291386</td>\n",
       "      <td>pos_control_mosaic_HB_HBE1_tss_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086491</th>\n",
       "      <td>chr11</td>\n",
       "      <td>5291385</td>\n",
       "      <td>5291386</td>\n",
       "      <td>pos_control_Klannchr1_HS4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gRNAgroup.chr gRNAgroup.start gRNAgroup.stop  \\\n",
       "1086199         chr11         5291385        5291386   \n",
       "1086272         chr11         5291385        5291386   \n",
       "1086345         chr11         5291385        5291386   \n",
       "1086418         chr11         5291385        5291386   \n",
       "1086491         chr11         5291385        5291386   \n",
       "\n",
       "                                        gRNAgroup  \n",
       "1086199              pos_control_HS2_Klann_mosaic  \n",
       "1086272  pos_control_Klannchr1_HBG1_HBG1_tss_both  \n",
       "1086345          pos_control_mosaic_HB_HBE1_tss_B  \n",
       "1086418          pos_control_mosaic_HB_HBE1_tss_A  \n",
       "1086491                 pos_control_Klannchr1_HS4  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intended_targets_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6f093fd9-8e3e-48fe-b0bf-ab3564bbafff",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['genomic_coords'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_23216/1718388157.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mvalidate_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgroup_sequence_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mintended_targets_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleft_on\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'gRNA_group'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mright_on\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'gRNAgroup'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m               \u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'genomic_coords'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gRNAgroup'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mvalidate_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py39/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4306\u001b[0m                 \u001b[0mweight\u001b[0m  \u001b[0;36m1.0\u001b[0m     \u001b[0;36m0.8\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4307\u001b[0m         \"\"\"\n\u001b[0;32m-> 4308\u001b[0;31m         return super().drop(\n\u001b[0m\u001b[1;32m   4309\u001b[0m             \u001b[0mlabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4310\u001b[0m             \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py39/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4151\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4152\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4153\u001b[0;31m                 \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_drop_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4155\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py39/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_drop_axis\u001b[0;34m(self, labels, axis, level, errors)\u001b[0m\n\u001b[1;32m   4186\u001b[0m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4187\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4188\u001b[0;31m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4189\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0maxis_name\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mnew_axis\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4190\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py39/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   5589\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5590\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merrors\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5591\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{labels[mask]} not found in axis\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5592\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5593\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['genomic_coords'] not found in axis\""
     ]
    }
   ],
   "source": [
    "validate_df = group_sequence_df.merge(intended_targets_df, left_on='gRNA_group', right_on='gRNAgroup')\\\n",
    "              .drop(['genomic_coords', 'gRNAgroup'], axis=1)\n",
    "validate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d99ec41d-e106-4c92-a16d-f44c7d6153cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13163, 9)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validate_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af990018-6104-4dcb-9e0d-82ac16a69b82",
   "metadata": {},
   "source": [
    "### Missingness Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e71f7936-f6c6-4e38-8eab-416bd4d3e72d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gRNA_group          0\n",
       "spacer_sequence     0\n",
       "genomic_coords      0\n",
       "start              12\n",
       "end                12\n",
       "chrom              12\n",
       "evalue             12\n",
       "dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get count for how many null values there are in the results\n",
    "group_sequence_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cc209f82-8263-4ef7-b243-75b792ef9321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>chr10.129_top_two</td>\n",
       "      <td>AGACCACACAACACACACAG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>chr10.2135_top_two</td>\n",
       "      <td>AGAGGTGTGTGTGTGTCACG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4048</th>\n",
       "      <td>chr15.931_top_two</td>\n",
       "      <td>CTGTTCTCTCTCTCTCTGCC</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6746</th>\n",
       "      <td>chr2.1666_top_two</td>\n",
       "      <td>TCCTGGACACACACACACAA</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13097</th>\n",
       "      <td>scrambled_10</td>\n",
       "      <td>ACTGCCTCGCGATTGACTGG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13102</th>\n",
       "      <td>scrambled_1</td>\n",
       "      <td>AGCCTAACGATCGGACCGAG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13121</th>\n",
       "      <td>scrambled_14</td>\n",
       "      <td>CAGGATCGCTATCAGCACGG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13149</th>\n",
       "      <td>scrambled_11</td>\n",
       "      <td>GCGACATTTGGGTCGCGAAG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13154</th>\n",
       "      <td>scrambled_8</td>\n",
       "      <td>GCTGTATATCGGCGCCCCGG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13155</th>\n",
       "      <td>scrambled_17</td>\n",
       "      <td>GGACGAGTAACCTGCCGGGG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13156</th>\n",
       "      <td>scrambled_16</td>\n",
       "      <td>GGCATCGGGCAGACGGGATG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13167</th>\n",
       "      <td>scrambled_12</td>\n",
       "      <td>TAAGGGCGACGTTCGGTAAG</td>\n",
       "      <td>[nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gRNA_group       spacer_sequence        genomic_coords  start  \\\n",
       "818     chr10.129_top_two  AGACCACACAACACACACAG  [nan, nan, nan, nan]    NaN   \n",
       "925    chr10.2135_top_two  AGAGGTGTGTGTGTGTCACG  [nan, nan, nan, nan]    NaN   \n",
       "4048    chr15.931_top_two  CTGTTCTCTCTCTCTCTGCC  [nan, nan, nan, nan]    NaN   \n",
       "6746    chr2.1666_top_two  TCCTGGACACACACACACAA  [nan, nan, nan, nan]    NaN   \n",
       "13097        scrambled_10  ACTGCCTCGCGATTGACTGG  [nan, nan, nan, nan]    NaN   \n",
       "13102         scrambled_1  AGCCTAACGATCGGACCGAG  [nan, nan, nan, nan]    NaN   \n",
       "13121        scrambled_14  CAGGATCGCTATCAGCACGG  [nan, nan, nan, nan]    NaN   \n",
       "13149        scrambled_11  GCGACATTTGGGTCGCGAAG  [nan, nan, nan, nan]    NaN   \n",
       "13154         scrambled_8  GCTGTATATCGGCGCCCCGG  [nan, nan, nan, nan]    NaN   \n",
       "13155        scrambled_17  GGACGAGTAACCTGCCGGGG  [nan, nan, nan, nan]    NaN   \n",
       "13156        scrambled_16  GGCATCGGGCAGACGGGATG  [nan, nan, nan, nan]    NaN   \n",
       "13167        scrambled_12  TAAGGGCGACGTTCGGTAAG  [nan, nan, nan, nan]    NaN   \n",
       "\n",
       "       end chrom  evalue  \n",
       "818    NaN   NaN     NaN  \n",
       "925    NaN   NaN     NaN  \n",
       "4048   NaN   NaN     NaN  \n",
       "6746   NaN   NaN     NaN  \n",
       "13097  NaN   NaN     NaN  \n",
       "13102  NaN   NaN     NaN  \n",
       "13121  NaN   NaN     NaN  \n",
       "13149  NaN   NaN     NaN  \n",
       "13154  NaN   NaN     NaN  \n",
       "13155  NaN   NaN     NaN  \n",
       "13156  NaN   NaN     NaN  \n",
       "13167  NaN   NaN     NaN  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# where the null values are coming from \n",
    "group_sequence_df[group_sequence_df['start'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "9907d3f2-5ea4-490e-8f53-f2dcf230d59a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10.129_top_two</td>\n",
       "      <td>AGACCACACAACACACACAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10</td>\n",
       "      <td>3710457</td>\n",
       "      <td>3711106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10.2135_top_two</td>\n",
       "      <td>AGAGGTGTGTGTGTGTCACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr10</td>\n",
       "      <td>71494457</td>\n",
       "      <td>71495376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr15.931_top_two</td>\n",
       "      <td>CTGTTCTCTCTCTCTCTGCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr15</td>\n",
       "      <td>45076438</td>\n",
       "      <td>45076745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2.1666_top_two</td>\n",
       "      <td>TCCTGGACACACACACACAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>43155472</td>\n",
       "      <td>43155535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scrambled_10</td>\n",
       "      <td>ACTGCCTCGCGATTGACTGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>scrambled_1</td>\n",
       "      <td>AGCCTAACGATCGGACCGAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>scrambled_14</td>\n",
       "      <td>CAGGATCGCTATCAGCACGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>scrambled_11</td>\n",
       "      <td>GCGACATTTGGGTCGCGAAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>scrambled_8</td>\n",
       "      <td>GCTGTATATCGGCGCCCCGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>scrambled_17</td>\n",
       "      <td>GGACGAGTAACCTGCCGGGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>scrambled_16</td>\n",
       "      <td>GGCATCGGGCAGACGGGATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>scrambled_12</td>\n",
       "      <td>TAAGGGCGACGTTCGGTAAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gRNA_group       spacer_sequence  start  end chrom  evalue  \\\n",
       "0    chr10.129_top_two  AGACCACACAACACACACAG    NaN  NaN   NaN     NaN   \n",
       "1   chr10.2135_top_two  AGAGGTGTGTGTGTGTCACG    NaN  NaN   NaN     NaN   \n",
       "2    chr15.931_top_two  CTGTTCTCTCTCTCTCTGCC    NaN  NaN   NaN     NaN   \n",
       "3    chr2.1666_top_two  TCCTGGACACACACACACAA    NaN  NaN   NaN     NaN   \n",
       "4         scrambled_10  ACTGCCTCGCGATTGACTGG    NaN  NaN   NaN     NaN   \n",
       "5          scrambled_1  AGCCTAACGATCGGACCGAG    NaN  NaN   NaN     NaN   \n",
       "6         scrambled_14  CAGGATCGCTATCAGCACGG    NaN  NaN   NaN     NaN   \n",
       "7         scrambled_11  GCGACATTTGGGTCGCGAAG    NaN  NaN   NaN     NaN   \n",
       "8          scrambled_8  GCTGTATATCGGCGCCCCGG    NaN  NaN   NaN     NaN   \n",
       "9         scrambled_17  GGACGAGTAACCTGCCGGGG    NaN  NaN   NaN     NaN   \n",
       "10        scrambled_16  GGCATCGGGCAGACGGGATG    NaN  NaN   NaN     NaN   \n",
       "11        scrambled_12  TAAGGGCGACGTTCGGTAAG    NaN  NaN   NaN     NaN   \n",
       "\n",
       "   gRNAgroup.chr gRNAgroup.start gRNAgroup.stop  \n",
       "0          chr10         3710457        3711106  \n",
       "1          chr10        71494457       71495376  \n",
       "2          chr15        45076438       45076745  \n",
       "3           chr2        43155472       43155535  \n",
       "4            NTC             NTC            NTC  \n",
       "5            NTC             NTC            NTC  \n",
       "6            NTC             NTC            NTC  \n",
       "7            NTC             NTC            NTC  \n",
       "8            NTC             NTC            NTC  \n",
       "9            NTC             NTC            NTC  \n",
       "10           NTC             NTC            NTC  \n",
       "11           NTC             NTC            NTC  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save dataframe and merge with intended targets\n",
    "null_coords_df = group_sequence_df[group_sequence_df['start'].isna()]\n",
    "null_coords_df.merge(intended_targets_df, left_on='gRNA_group', right_on='gRNAgroup')\\\n",
    "                     .drop(['gRNAgroup', 'genomic_coords'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "abbf5c01-2d18-4305-90f6-2b5c38ef069b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13062</th>\n",
       "      <td>random_24</td>\n",
       "      <td>AACACAACACACCAAAACTG</td>\n",
       "      <td>101428.0</td>\n",
       "      <td>101409.0</td>\n",
       "      <td>chr9_gl000200_random</td>\n",
       "      <td>0.008</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13063</th>\n",
       "      <td>random_24</td>\n",
       "      <td>GCAAATGCTTCATCACCCCA</td>\n",
       "      <td>11074158.0</td>\n",
       "      <td>11074139.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13064</th>\n",
       "      <td>random_7</td>\n",
       "      <td>AAGTTGACTCTACATAGCAG</td>\n",
       "      <td>23770078.0</td>\n",
       "      <td>23770097.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13065</th>\n",
       "      <td>random_7</td>\n",
       "      <td>GCTCTAATGAACAGAATGGG</td>\n",
       "      <td>25699943.0</td>\n",
       "      <td>25699924.0</td>\n",
       "      <td>chr4</td>\n",
       "      <td>0.008</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13066</th>\n",
       "      <td>random_13</td>\n",
       "      <td>AATATTCTCCCTCATTCTGG</td>\n",
       "      <td>12539494.0</td>\n",
       "      <td>12539475.0</td>\n",
       "      <td>chr5</td>\n",
       "      <td>0.008</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gRNA_group       spacer_sequence       start         end  \\\n",
       "13062  random_24  AACACAACACACCAAAACTG    101428.0    101409.0   \n",
       "13063  random_24  GCAAATGCTTCATCACCCCA  11074158.0  11074139.0   \n",
       "13064   random_7  AAGTTGACTCTACATAGCAG  23770078.0  23770097.0   \n",
       "13065   random_7  GCTCTAATGAACAGAATGGG  25699943.0  25699924.0   \n",
       "13066  random_13  AATATTCTCCCTCATTCTGG  12539494.0  12539475.0   \n",
       "\n",
       "                      chrom  evalue gRNAgroup.chr gRNAgroup.start  \\\n",
       "13062  chr9_gl000200_random   0.008           NTC             NTC   \n",
       "13063                  chr8   0.008           NTC             NTC   \n",
       "13064                  chr8   0.008           NTC             NTC   \n",
       "13065                  chr4   0.008           NTC             NTC   \n",
       "13066                  chr5   0.008           NTC             NTC   \n",
       "\n",
       "      gRNAgroup.stop  \n",
       "13062            NTC  \n",
       "13063            NTC  \n",
       "13064            NTC  \n",
       "13065            NTC  \n",
       "13066            NTC  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# view NTC groups from original sequences\n",
    "group_sequence_targets = group_sequence_df.merge(intended_targets_df, left_on='gRNA_group', right_on='gRNAgroup')\n",
    "group_sequence_targets[group_sequence_targets['gRNAgroup.chr'] == 'NTC'].drop(['gRNAgroup', 'genomic_coords'], axis=1).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0ab647b3-11cc-4d2d-be17-a5d6a9b8ad3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(101, 11)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_sequence_targets[group_sequence_targets['gRNAgroup.chr'] == 'NTC'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ef200735-fc35-43b3-adbf-5cd6d1c24825",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13062</th>\n",
       "      <td>chr9_gl000200_random</td>\n",
       "      <td>101428</td>\n",
       "      <td>101409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13063</th>\n",
       "      <td>chr8</td>\n",
       "      <td>11074158</td>\n",
       "      <td>11074139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13064</th>\n",
       "      <td>chr8</td>\n",
       "      <td>23770078</td>\n",
       "      <td>23770097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13065</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25699943</td>\n",
       "      <td>25699924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13066</th>\n",
       "      <td>chr5</td>\n",
       "      <td>12539494</td>\n",
       "      <td>12539475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      chrom     start       end\n",
       "13062  chr9_gl000200_random    101428    101409\n",
       "13063                  chr8  11074158  11074139\n",
       "13064                  chr8  23770078  23770097\n",
       "13065                  chr4  25699943  25699924\n",
       "13066                  chr5  12539494  12539475"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntc_df = group_sequence_targets[group_sequence_targets['gRNAgroup.chr'] == 'NTC']\n",
    "ntc_df = ntc_df[['chrom', 'start', 'end']].dropna()\n",
    "ntc_df['start'] = ntc_df['start'].astype(np.int64)\n",
    "ntc_df['end'] = ntc_df['end'].astype(np.int64)\n",
    "ntc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c773c797-b62c-4d2e-8657-e41e882f5844",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write non targeting controls with hits to .bed file (for browser viewing)\n",
    "ntc_df.to_csv('./../data/ntc_blast_hits.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65823ea2-089b-4742-af96-eb4c5e581525",
   "metadata": {},
   "source": [
    "There are 101 nontargeting controls, but most of them return coordinates from BLAST! The evalues appear to be pretty low for these hits as well, indicating potential off-target effects?\n",
    "\n",
    "There also appears to be 4 guides that do not get a BLAST hit despite having intended target regions. This could be indicative of low guide specificity?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "414fec9c-3afd-4ca0-a4b5-802ac0c4e4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save BLAST output to .csv file\n",
    "group_sequence_targets.to_csv('./../data/group_sequence_coords.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a82c7e97-ef88-4a4e-8529-aeae0a930fc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>[26606551, 26606569, 'chr1', 0.032]</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>[26606660, 26606641, 'chr1', 0.008]</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>[10530445, 10530427, 'chr11', 0.032]</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>[10530690, 10530708, 'chr11', 0.032]</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>[124195027, 124195008, 'chr8', 0.008]</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr8</td>\n",
       "      <td>124191287</td>\n",
       "      <td>124191288</td>\n",
       "      <td>FAM83A_TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence                         genomic_coords  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG    [26606551, 26606569, 'chr1', 0.032]   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG    [26606660, 26606641, 'chr1', 0.008]   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   [10530445, 10530427, 'chr11', 0.032]   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   [10530690, 10530708, 'chr11', 0.032]   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  [124195027, 124195008, 'chr8', 0.008]   \n",
       "\n",
       "         start          end  chrom  evalue gRNAgroup.chr gRNAgroup.start  \\\n",
       "0   26606551.0   26606569.0   chr1   0.032          chr1        26605667   \n",
       "1   26606660.0   26606641.0   chr1   0.008          chr1        26605667   \n",
       "2   10530445.0   10530427.0  chr11   0.032         chr11        10530735   \n",
       "3   10530690.0   10530708.0  chr11   0.032         chr11        10530735   \n",
       "4  124195027.0  124195008.0   chr8   0.008          chr8       124191287   \n",
       "\n",
       "  gRNAgroup.stop     gRNAgroup  \n",
       "0       26605668  SH3BGRL3_TSS  \n",
       "1       26605668  SH3BGRL3_TSS  \n",
       "2       10530736  MTRNR2L8_TSS  \n",
       "3       10530736  MTRNR2L8_TSS  \n",
       "4      124191288    FAM83A_TSS  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alternative run to load in resulting data frame from CSV\n",
    "group_sequence_targets = pd.read_csv('./../data/group_sequence_coords.csv')\n",
    "group_sequence_targets.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4db92a8b-af38-4590-b3a2-775ba422b9c9",
   "metadata": {},
   "source": [
    "### Viewing Count Data Using scipy.sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15e80fc7-bf30-48c9-8159-83d5522f8370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13135, 207324)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in count data\n",
    "count_matrix = mmread('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.exprs.mtx')\n",
    "count_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "761c3320-94ad-467d-80a4-2c4c55dfcb5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2\n",
       "1         AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2\n",
       "2         AAACCTGCAAACAACA-1_1A_1_SI-GA-E2\n",
       "3         AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2\n",
       "4         AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2\n",
       "                        ...               \n",
       "207319    TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9\n",
       "207320    TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9\n",
       "207321    TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9\n",
       "207322    TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9\n",
       "207323    TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9\n",
       "Length: 207324, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in column names from corresponding cells file\n",
    "colnames = []\n",
    "with open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.cells.txt') as f:\n",
    "    colnames = f.readlines()\n",
    "\n",
    "colnames = pd.Series(colnames).str.strip()\n",
    "colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "77ffdaa8-1a1d-4369-a0c6-c6bd75d1b377",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        ENSG00000238009\n",
       "1        ENSG00000237683\n",
       "2        ENSG00000228463\n",
       "3        ENSG00000237094\n",
       "4        ENSG00000235373\n",
       "              ...       \n",
       "13130    ENSG00000215689\n",
       "13131    ENSG00000215781\n",
       "13132    ENSG00000220023\n",
       "13133    ENSG00000215615\n",
       "13134    ENSG00000215699\n",
       "Length: 13135, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in index (row names) from corresponding genes file \n",
    "rownames = []\n",
    "with open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.genes.txt') as f:\n",
    "    rownames = f.readlines()\n",
    "    \n",
    "rownames = pd.Series(rownames).str.strip()\n",
    "rownames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b92ed7bc-2b86-4cf6-80b5-73693083ae63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTATATGAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACAACGATGG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGATAATG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237683</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237094</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000235373</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228327</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237491</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000225880</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000230368</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188157</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000131591</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000078808</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000176022</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000184163</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160087</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows  207324 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 1   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 1   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 4   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 1   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 1   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 3   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 2   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 1   \n",
       "ENSG00000160087                                 1   \n",
       "\n",
       "                 AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 1   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 1   \n",
       "ENSG00000187608                                 2   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 1   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 1   \n",
       "ENSG00000160087                                 3   \n",
       "\n",
       "                 AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 1   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 1   \n",
       "\n",
       "                 AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 1   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 1   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 1   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 1   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 AAACCTGGTATATGAG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 1   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 1   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 2   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2  ...  \\\n",
       "ENSG00000238009                                 0  ...   \n",
       "ENSG00000237683                                 0  ...   \n",
       "ENSG00000228463                                 1  ...   \n",
       "ENSG00000237094                                 0  ...   \n",
       "ENSG00000235373                                 0  ...   \n",
       "ENSG00000228327                                 0  ...   \n",
       "ENSG00000237491                                 0  ...   \n",
       "ENSG00000225880                                 0  ...   \n",
       "ENSG00000230368                                 0  ...   \n",
       "ENSG00000188976                                 1  ...   \n",
       "ENSG00000187961                                 0  ...   \n",
       "ENSG00000187583                                 0  ...   \n",
       "ENSG00000188290                                 0  ...   \n",
       "ENSG00000187608                                 2  ...   \n",
       "ENSG00000188157                                 0  ...   \n",
       "ENSG00000131591                                 0  ...   \n",
       "ENSG00000078808                                 1  ...   \n",
       "ENSG00000176022                                 0  ...   \n",
       "ENSG00000184163                                 0  ...   \n",
       "ENSG00000160087                                 3  ...   \n",
       "\n",
       "                 TTTGTCACAACGATGG-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 2   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 4   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 1   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 4   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 TTTGTCACAGATAATG-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 2   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 1   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 2   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 1   \n",
       "\n",
       "                 TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 0   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 2   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 2   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 1   \n",
       "ENSG00000176022                                 1   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 2   \n",
       "\n",
       "                 TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 3   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 1   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 1   \n",
       "ENSG00000176022                                 1   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 3   \n",
       "\n",
       "                 TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 2   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 1   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 4   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 0   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 1   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 1   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "ENSG00000228327                                 0   \n",
       "ENSG00000237491                                 0   \n",
       "ENSG00000225880                                 0   \n",
       "ENSG00000230368                                 0   \n",
       "ENSG00000188976                                 1   \n",
       "ENSG00000187961                                 0   \n",
       "ENSG00000187583                                 0   \n",
       "ENSG00000188290                                 1   \n",
       "ENSG00000187608                                 0   \n",
       "ENSG00000188157                                 0   \n",
       "ENSG00000131591                                 0   \n",
       "ENSG00000078808                                 0   \n",
       "ENSG00000176022                                 0   \n",
       "ENSG00000184163                                 0   \n",
       "ENSG00000160087                                 0   \n",
       "\n",
       "                 TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  \n",
       "ENSG00000238009                                 0  \n",
       "ENSG00000237683                                 0  \n",
       "ENSG00000228463                                 2  \n",
       "ENSG00000237094                                 0  \n",
       "ENSG00000235373                                 0  \n",
       "ENSG00000228327                                 0  \n",
       "ENSG00000237491                                 0  \n",
       "ENSG00000225880                                 0  \n",
       "ENSG00000230368                                 0  \n",
       "ENSG00000188976                                 2  \n",
       "ENSG00000187961                                 0  \n",
       "ENSG00000187583                                 0  \n",
       "ENSG00000188290                                 0  \n",
       "ENSG00000187608                                 2  \n",
       "ENSG00000188157                                 0  \n",
       "ENSG00000131591                                 0  \n",
       "ENSG00000078808                                 2  \n",
       "ENSG00000176022                                 0  \n",
       "ENSG00000184163                                 0  \n",
       "ENSG00000160087                                 1  \n",
       "\n",
       "[20 rows x 207324 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create formatted data frame with row and column names\n",
    "count_data = pd.DataFrame(data=count_matrix.toarray(), index=rownames, columns=colnames)\n",
    "count_data.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "900643fb-215d-43d4-81fd-30bd0c839bcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2    11\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2     2\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     9\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2    11\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     3\n",
       "                                    ..\n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9     7\n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9     9\n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    10\n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9     2\n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9     9\n",
       "Length: 207324, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_subset = count_data.head(20)\n",
    "count_subset.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e576afb9-6dc8-4ba5-baff-086c0f0a9785",
   "metadata": {},
   "source": [
    "### Generate Cell-Guide Matrix (which cells contain which guides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a966f5e0-8d35-4566-acda-3185510a07ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>gene</th>\n",
       "      <th>all_gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>sample_directory</th>\n",
       "      <th>ko_barcode_file</th>\n",
       "      <th>id</th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>within_batch_chip</th>\n",
       "      <th>within_chip_lane</th>\n",
       "      <th>percent.mito</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>17572.0</td>\n",
       "      <td>1.009682</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...</td>\n",
       "      <td>14135.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.969819</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.058787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>8923.0</td>\n",
       "      <td>0.939677</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...</td>\n",
       "      <td>4329.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.844380</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.036087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>14637.0</td>\n",
       "      <td>0.990803</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...</td>\n",
       "      <td>12362.0</td>\n",
       "      <td>884.0</td>\n",
       "      <td>0.950538</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.069823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>22798.0</td>\n",
       "      <td>1.036578</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...</td>\n",
       "      <td>7459.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>0.939551</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.026187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>10136.0</td>\n",
       "      <td>0.952844</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...</td>\n",
       "      <td>14831.0</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>0.959927</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.007991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample                              cell  total_umis  Size_Factor  \\\n",
       "0  1A_1_SI-GA-E2  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2     17572.0     1.009682   \n",
       "1  1A_1_SI-GA-E2  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2      8923.0     0.939677   \n",
       "2  1A_1_SI-GA-E2  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     14637.0     0.990803   \n",
       "3  1A_1_SI-GA-E2  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2     22798.0     1.036578   \n",
       "4  1A_1_SI-GA-E2  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     10136.0     0.952844   \n",
       "\n",
       "                                                gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                            all_gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                             barcode  read_count  umi_count  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...     14135.0      964.0   \n",
       "1  GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...      4329.0      293.0   \n",
       "2  CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...     12362.0      884.0   \n",
       "3  GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...      7459.0      544.0   \n",
       "4  ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...     14831.0     1054.0   \n",
       "\n",
       "   proportion  guide_count sample_directory  \\\n",
       "0    0.969819         67.0    1A_1_SI-GA-E2   \n",
       "1    0.844380         26.0    1A_1_SI-GA-E2   \n",
       "2    0.950538         61.0    1A_1_SI-GA-E2   \n",
       "3    0.939551         39.0    1A_1_SI-GA-E2   \n",
       "4    0.959927         37.0    1A_1_SI-GA-E2   \n",
       "\n",
       "                         ko_barcode_file    id    prep_batch  \\\n",
       "0  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "1  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "2  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "3  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "4  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "\n",
       "     within_batch_chip    within_chip_lane  percent.mito  \n",
       "0  within_batch_chip_A  within_chip_lane_1      0.058787  \n",
       "1  within_batch_chip_A  within_chip_lane_1      0.036087  \n",
       "2  within_batch_chip_A  within_chip_lane_1      0.069823  \n",
       "3  within_batch_chip_A  within_chip_lane_1      0.026187  \n",
       "4  within_batch_chip_A  within_chip_lane_1      0.007991  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in phenodata file \n",
    "colnames = open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale.phenoData.colnames.txt') \\\n",
    "           .read().splitlines()\n",
    "\n",
    "with gzip.open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.phenoData.txt.gz', 'rt') as file:\n",
    "    phenodata_file = file.read()\n",
    "\n",
    "with open('./../data/phenodata.txt', 'w') as file:\n",
    "    file.write(phenodata_file)\n",
    "    \n",
    "phenodata_df = pd.read_csv('./../data/phenodata.txt', sep=' ', names=colnames)\n",
    "phenodata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "99ecf5c2-17bb-431a-9c0e-558adb22cfa0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(207324, 18)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print shape of phenodata data frame\n",
    "phenodata_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c5a91f9-b62c-4aaa-86dc-8042a978a06d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample                  0\n",
       "cell                    0\n",
       "total_umis              0\n",
       "Size_Factor             0\n",
       "gene                 1527\n",
       "all_gene             1527\n",
       "barcode              1527\n",
       "read_count           1527\n",
       "umi_count            1527\n",
       "proportion           1527\n",
       "guide_count          1527\n",
       "sample_directory        0\n",
       "ko_barcode_file         0\n",
       "id                      0\n",
       "prep_batch              0\n",
       "within_batch_chip       0\n",
       "within_chip_lane        0\n",
       "percent.mito            0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check dataframe missingness\n",
    "phenodata_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "828567f0-e7c2-4648-bf1e-e0d0531bc1cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>gene</th>\n",
       "      <th>all_gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>sample_directory</th>\n",
       "      <th>ko_barcode_file</th>\n",
       "      <th>id</th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>within_batch_chip</th>\n",
       "      <th>within_chip_lane</th>\n",
       "      <th>percent.mito</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAAGCAATCCTACAGA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>3971.0</td>\n",
       "      <td>0.856042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.016117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1012</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AGCGTATCAAACCCAT-1_1A_1_SI-GA-E2</td>\n",
       "      <td>5090.0</td>\n",
       "      <td>0.881687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.027112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AGCGTCGTCAGGTAAA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>5587.0</td>\n",
       "      <td>0.891312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.017720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AGCGTCGTCGCGATCG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>16844.0</td>\n",
       "      <td>1.005311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.061149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AGTAGTCTCAGGATCT-1_1A_1_SI-GA-E2</td>\n",
       "      <td>4958.0</td>\n",
       "      <td>0.878973</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.041549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207126</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>TTCTACATCGAGCCCA-1_2B_8_SI-GA-H9</td>\n",
       "      <td>5833.0</td>\n",
       "      <td>0.895763</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>guide_libraries/2B_8.gRNAcaptured.txt</td>\n",
       "      <td>2B_8</td>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "      <td>0.013544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207187</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>TTGCCGTGTAGTGAAT-1_2B_8_SI-GA-H9</td>\n",
       "      <td>5955.0</td>\n",
       "      <td>0.897901</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>guide_libraries/2B_8.gRNAcaptured.txt</td>\n",
       "      <td>2B_8</td>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "      <td>0.051889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207202</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>TTGCGTCGTGAGTGAC-1_2B_8_SI-GA-H9</td>\n",
       "      <td>10269.0</td>\n",
       "      <td>0.954190</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>guide_libraries/2B_8.gRNAcaptured.txt</td>\n",
       "      <td>2B_8</td>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "      <td>0.028727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207205</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>TTGCGTCTCGTATCAG-1_2B_8_SI-GA-H9</td>\n",
       "      <td>7367.0</td>\n",
       "      <td>0.919882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>guide_libraries/2B_8.gRNAcaptured.txt</td>\n",
       "      <td>2B_8</td>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "      <td>0.004344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207290</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>TTTGCGCAGATCCCAT-1_2B_8_SI-GA-H9</td>\n",
       "      <td>5225.0</td>\n",
       "      <td>0.884392</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>guide_libraries/2B_8.gRNAcaptured.txt</td>\n",
       "      <td>2B_8</td>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "      <td>0.014928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1527 rows  18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample                              cell  total_umis  \\\n",
       "49      1A_1_SI-GA-E2  AAAGCAATCCTACAGA-1_1A_1_SI-GA-E2      3971.0   \n",
       "1012    1A_1_SI-GA-E2  AGCGTATCAAACCCAT-1_1A_1_SI-GA-E2      5090.0   \n",
       "1031    1A_1_SI-GA-E2  AGCGTCGTCAGGTAAA-1_1A_1_SI-GA-E2      5587.0   \n",
       "1034    1A_1_SI-GA-E2  AGCGTCGTCGCGATCG-1_1A_1_SI-GA-E2     16844.0   \n",
       "1194    1A_1_SI-GA-E2  AGTAGTCTCAGGATCT-1_1A_1_SI-GA-E2      4958.0   \n",
       "...               ...                               ...         ...   \n",
       "207126  2B_8_SI-GA-H9  TTCTACATCGAGCCCA-1_2B_8_SI-GA-H9      5833.0   \n",
       "207187  2B_8_SI-GA-H9  TTGCCGTGTAGTGAAT-1_2B_8_SI-GA-H9      5955.0   \n",
       "207202  2B_8_SI-GA-H9  TTGCGTCGTGAGTGAC-1_2B_8_SI-GA-H9     10269.0   \n",
       "207205  2B_8_SI-GA-H9  TTGCGTCTCGTATCAG-1_2B_8_SI-GA-H9      7367.0   \n",
       "207290  2B_8_SI-GA-H9  TTTGCGCAGATCCCAT-1_2B_8_SI-GA-H9      5225.0   \n",
       "\n",
       "        Size_Factor gene all_gene barcode  read_count  umi_count  proportion  \\\n",
       "49         0.856042  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "1012       0.881687  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "1031       0.891312  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "1034       1.005311  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "1194       0.878973  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "...             ...  ...      ...     ...         ...        ...         ...   \n",
       "207126     0.895763  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "207187     0.897901  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "207202     0.954190  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "207205     0.919882  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "207290     0.884392  NaN      NaN     NaN         NaN        NaN         NaN   \n",
       "\n",
       "        guide_count sample_directory                        ko_barcode_file  \\\n",
       "49              NaN    1A_1_SI-GA-E2  guide_libraries/1A_1.gRNAcaptured.txt   \n",
       "1012            NaN    1A_1_SI-GA-E2  guide_libraries/1A_1.gRNAcaptured.txt   \n",
       "1031            NaN    1A_1_SI-GA-E2  guide_libraries/1A_1.gRNAcaptured.txt   \n",
       "1034            NaN    1A_1_SI-GA-E2  guide_libraries/1A_1.gRNAcaptured.txt   \n",
       "1194            NaN    1A_1_SI-GA-E2  guide_libraries/1A_1.gRNAcaptured.txt   \n",
       "...             ...              ...                                    ...   \n",
       "207126          NaN    2B_8_SI-GA-H9  guide_libraries/2B_8.gRNAcaptured.txt   \n",
       "207187          NaN    2B_8_SI-GA-H9  guide_libraries/2B_8.gRNAcaptured.txt   \n",
       "207202          NaN    2B_8_SI-GA-H9  guide_libraries/2B_8.gRNAcaptured.txt   \n",
       "207205          NaN    2B_8_SI-GA-H9  guide_libraries/2B_8.gRNAcaptured.txt   \n",
       "207290          NaN    2B_8_SI-GA-H9  guide_libraries/2B_8.gRNAcaptured.txt   \n",
       "\n",
       "          id    prep_batch    within_batch_chip    within_chip_lane  \\\n",
       "49      1A_1  prep_batch_1  within_batch_chip_A  within_chip_lane_1   \n",
       "1012    1A_1  prep_batch_1  within_batch_chip_A  within_chip_lane_1   \n",
       "1031    1A_1  prep_batch_1  within_batch_chip_A  within_chip_lane_1   \n",
       "1034    1A_1  prep_batch_1  within_batch_chip_A  within_chip_lane_1   \n",
       "1194    1A_1  prep_batch_1  within_batch_chip_A  within_chip_lane_1   \n",
       "...      ...           ...                  ...                 ...   \n",
       "207126  2B_8  prep_batch_2  within_batch_chip_B  within_chip_lane_8   \n",
       "207187  2B_8  prep_batch_2  within_batch_chip_B  within_chip_lane_8   \n",
       "207202  2B_8  prep_batch_2  within_batch_chip_B  within_chip_lane_8   \n",
       "207205  2B_8  prep_batch_2  within_batch_chip_B  within_chip_lane_8   \n",
       "207290  2B_8  prep_batch_2  within_batch_chip_B  within_chip_lane_8   \n",
       "\n",
       "        percent.mito  \n",
       "49          0.016117  \n",
       "1012        0.027112  \n",
       "1031        0.017720  \n",
       "1034        0.061149  \n",
       "1194        0.041549  \n",
       "...              ...  \n",
       "207126      0.013544  \n",
       "207187      0.051889  \n",
       "207202      0.028727  \n",
       "207205      0.004344  \n",
       "207290      0.014928  \n",
       "\n",
       "[1527 rows x 18 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# analyze missingness\n",
    "phenodata_df[phenodata_df['gene'].apply(type) != str]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1d3ebe48-a9b7-40e1-bd34-c9d6ecbb93cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>gene</th>\n",
       "      <th>all_gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>sample_directory</th>\n",
       "      <th>ko_barcode_file</th>\n",
       "      <th>id</th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>within_batch_chip</th>\n",
       "      <th>within_chip_lane</th>\n",
       "      <th>percent.mito</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>17572.0</td>\n",
       "      <td>1.009682</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...</td>\n",
       "      <td>14135.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.969819</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.058787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>8923.0</td>\n",
       "      <td>0.939677</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...</td>\n",
       "      <td>4329.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.844380</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.036087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>14637.0</td>\n",
       "      <td>0.990803</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...</td>\n",
       "      <td>12362.0</td>\n",
       "      <td>884.0</td>\n",
       "      <td>0.950538</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.069823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>22798.0</td>\n",
       "      <td>1.036578</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...</td>\n",
       "      <td>7459.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>0.939551</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.026187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>10136.0</td>\n",
       "      <td>0.952844</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...</td>\n",
       "      <td>14831.0</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>0.959927</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.007991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample                              cell  total_umis  Size_Factor  \\\n",
       "0  1A_1_SI-GA-E2  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2     17572.0     1.009682   \n",
       "1  1A_1_SI-GA-E2  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2      8923.0     0.939677   \n",
       "2  1A_1_SI-GA-E2  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     14637.0     0.990803   \n",
       "3  1A_1_SI-GA-E2  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2     22798.0     1.036578   \n",
       "4  1A_1_SI-GA-E2  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     10136.0     0.952844   \n",
       "\n",
       "                                                gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                            all_gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                             barcode  read_count  umi_count  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...     14135.0      964.0   \n",
       "1  GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...      4329.0      293.0   \n",
       "2  CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...     12362.0      884.0   \n",
       "3  GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...      7459.0      544.0   \n",
       "4  ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...     14831.0     1054.0   \n",
       "\n",
       "   proportion  guide_count sample_directory  \\\n",
       "0    0.969819         67.0    1A_1_SI-GA-E2   \n",
       "1    0.844380         26.0    1A_1_SI-GA-E2   \n",
       "2    0.950538         61.0    1A_1_SI-GA-E2   \n",
       "3    0.939551         39.0    1A_1_SI-GA-E2   \n",
       "4    0.959927         37.0    1A_1_SI-GA-E2   \n",
       "\n",
       "                         ko_barcode_file    id    prep_batch  \\\n",
       "0  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "1  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "2  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "3  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "4  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "\n",
       "     within_batch_chip    within_chip_lane  percent.mito  \n",
       "0  within_batch_chip_A  within_chip_lane_1      0.058787  \n",
       "1  within_batch_chip_A  within_chip_lane_1      0.036087  \n",
       "2  within_batch_chip_A  within_chip_lane_1      0.069823  \n",
       "3  within_batch_chip_A  within_chip_lane_1      0.026187  \n",
       "4  within_batch_chip_A  within_chip_lane_1      0.007991  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop nan values fromd data frame\n",
    "phenodata_df = phenodata_df.dropna()\n",
    "phenodata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "611ba399-57be-438a-9d1e-0df7d12e227c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(205797, 18)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get new data frame shape\n",
    "phenodata_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "09838e5c-d380-4989-9f19-9e5096be49f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_guide_names(guides):\n",
    "    '''regex function to separate guide names'''\n",
    "    \n",
    "    guide_list = re.findall(r\"[A-Za-z0-9]*\\_TSS|chr[A-Z0-9]{1,2}\\.\\w*top_two|chr[A-Z0-9]{1,2}\\.\\w*second_two\", guides)\n",
    "    return guide_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fc9d3818-66b7-41c1-a73f-66e0f416f2ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    67\n",
       "1    26\n",
       "2    61\n",
       "3    39\n",
       "4    37\n",
       "Name: guides, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get guide names for each cell\n",
    "phenodata_df['guides'] = phenodata_df['gene'].apply(get_guide_names)\n",
    "test = phenodata_df.head()\n",
    "test['guides'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c6f22619-e063-4bf0-9f33-fec8b3e5df71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>Size_Factor</th>\n",
       "      <th>gene</th>\n",
       "      <th>all_gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>sample_directory</th>\n",
       "      <th>ko_barcode_file</th>\n",
       "      <th>id</th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>within_batch_chip</th>\n",
       "      <th>within_chip_lane</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>guides</th>\n",
       "      <th>sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>17572.0</td>\n",
       "      <td>1.009682</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...</td>\n",
       "      <td>14135.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.969819</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.058787</td>\n",
       "      <td>[chr10.845_top_two, chr1.11183_top_two, chr1.1...</td>\n",
       "      <td>[AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>8923.0</td>\n",
       "      <td>0.939677</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...</td>\n",
       "      <td>4329.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.844380</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.036087</td>\n",
       "      <td>[chr1.12695_top_two, chr11.3294_top_two, chr1....</td>\n",
       "      <td>[GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>14637.0</td>\n",
       "      <td>0.990803</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...</td>\n",
       "      <td>12362.0</td>\n",
       "      <td>884.0</td>\n",
       "      <td>0.950538</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.069823</td>\n",
       "      <td>[ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...</td>\n",
       "      <td>[CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>22798.0</td>\n",
       "      <td>1.036578</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...</td>\n",
       "      <td>7459.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>0.939551</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.026187</td>\n",
       "      <td>[C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...</td>\n",
       "      <td>[GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>10136.0</td>\n",
       "      <td>0.952844</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...</td>\n",
       "      <td>14831.0</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>0.959927</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>guide_libraries/1A_1.gRNAcaptured.txt</td>\n",
       "      <td>1A_1</td>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "      <td>0.007991</td>\n",
       "      <td>[chr10.185_top_two, chr10.484_top_two, chr11.4...</td>\n",
       "      <td>[ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample                              cell  total_umis  Size_Factor  \\\n",
       "0  1A_1_SI-GA-E2  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2     17572.0     1.009682   \n",
       "1  1A_1_SI-GA-E2  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2      8923.0     0.939677   \n",
       "2  1A_1_SI-GA-E2  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     14637.0     0.990803   \n",
       "3  1A_1_SI-GA-E2  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2     22798.0     1.036578   \n",
       "4  1A_1_SI-GA-E2  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     10136.0     0.952844   \n",
       "\n",
       "                                                gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                            all_gene  \\\n",
       "0  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "1  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "3  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "4  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "\n",
       "                                             barcode  read_count  umi_count  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC_TGATCGCTTTGACTGTGACA_ACAA...     14135.0      964.0   \n",
       "1  GTAGAGCCTCCAGAACTGTG_AGGTTTATCCAGATGAACTG_CATC...      4329.0      293.0   \n",
       "2  CCAAGGCGTCCTCAGACCAG_AGCTCCAGGAAGGACCCCCG_TCAC...     12362.0      884.0   \n",
       "3  GGCGTCAGTCGAGGAGTCAG_GCCAGCACTTCAGCTCACCG_GCTG...      7459.0      544.0   \n",
       "4  ATAAGGCACTCACATCCACC_GCTTGTCCCTAACACTCAGA_GGGC...     14831.0     1054.0   \n",
       "\n",
       "   proportion  guide_count sample_directory  \\\n",
       "0    0.969819         67.0    1A_1_SI-GA-E2   \n",
       "1    0.844380         26.0    1A_1_SI-GA-E2   \n",
       "2    0.950538         61.0    1A_1_SI-GA-E2   \n",
       "3    0.939551         39.0    1A_1_SI-GA-E2   \n",
       "4    0.959927         37.0    1A_1_SI-GA-E2   \n",
       "\n",
       "                         ko_barcode_file    id    prep_batch  \\\n",
       "0  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "1  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "2  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "3  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "4  guide_libraries/1A_1.gRNAcaptured.txt  1A_1  prep_batch_1   \n",
       "\n",
       "     within_batch_chip    within_chip_lane  percent.mito  \\\n",
       "0  within_batch_chip_A  within_chip_lane_1      0.058787   \n",
       "1  within_batch_chip_A  within_chip_lane_1      0.036087   \n",
       "2  within_batch_chip_A  within_chip_lane_1      0.069823   \n",
       "3  within_batch_chip_A  within_chip_lane_1      0.026187   \n",
       "4  within_batch_chip_A  within_chip_lane_1      0.007991   \n",
       "\n",
       "                                              guides  \\\n",
       "0  [chr10.845_top_two, chr1.11183_top_two, chr1.1...   \n",
       "1  [chr1.12695_top_two, chr11.3294_top_two, chr1....   \n",
       "2  [ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...   \n",
       "3  [C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...   \n",
       "4  [chr10.185_top_two, chr10.484_top_two, chr11.4...   \n",
       "\n",
       "                                           sequences  \n",
       "0  [AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...  \n",
       "1  [GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...  \n",
       "2  [CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...  \n",
       "3  [GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...  \n",
       "4  [ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split guide sequences for each cell\n",
    "phenodata_df['sequences'] = phenodata_df['barcode'].str.split('_')\n",
    "phenodata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b3614411-f83d-4981-b4cc-8bbe47e77e14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>guides</th>\n",
       "      <th>sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr10.845_top_two, chr1.11183_top_two, chr1.1...</td>\n",
       "      <td>[AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr1.12695_top_two, chr11.3294_top_two, chr1....</td>\n",
       "      <td>[GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...</td>\n",
       "      <td>[CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...</td>\n",
       "      <td>[GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr10.185_top_two, chr10.484_top_two, chr11.4...</td>\n",
       "      <td>[ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cell  \\\n",
       "0  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2   \n",
       "1  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2   \n",
       "2  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2   \n",
       "3  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2   \n",
       "4  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2   \n",
       "\n",
       "                                              guides  \\\n",
       "0  [chr10.845_top_two, chr1.11183_top_two, chr1.1...   \n",
       "1  [chr1.12695_top_two, chr11.3294_top_two, chr1....   \n",
       "2  [ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...   \n",
       "3  [C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...   \n",
       "4  [chr10.185_top_two, chr10.484_top_two, chr11.4...   \n",
       "\n",
       "                                           sequences  \n",
       "0  [AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...  \n",
       "1  [GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...  \n",
       "2  [CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...  \n",
       "3  [GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...  \n",
       "4  [ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select only columns necessary for analysis\n",
    "phenodata_df = phenodata_df[['cell', 'guides', 'sequences']]\n",
    "phenodata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f930f110-5f32-40ef-bff7-95d13b0d6697",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>guides</th>\n",
       "      <th>sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr10.845_top_two, chr1.11183_top_two, chr1.1...</td>\n",
       "      <td>[AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr1.12695_top_two, chr11.3294_top_two, chr1....</td>\n",
       "      <td>[GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...</td>\n",
       "      <td>[CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...</td>\n",
       "      <td>[GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>[chr10.185_top_two, chr10.484_top_two, chr11.4...</td>\n",
       "      <td>[ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cell  \\\n",
       "0  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2   \n",
       "1  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2   \n",
       "2  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2   \n",
       "3  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2   \n",
       "4  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2   \n",
       "\n",
       "                                              guides  \\\n",
       "0  [chr10.845_top_two, chr1.11183_top_two, chr1.1...   \n",
       "1  [chr1.12695_top_two, chr11.3294_top_two, chr1....   \n",
       "2  [ALDH1A2_TSS, BRI3_TSS, chr10.1918_top_two, ch...   \n",
       "3  [C16orf91_TSS, chr1.11332_top_two, chr1.1933_t...   \n",
       "4  [chr10.185_top_two, chr10.484_top_two, chr11.4...   \n",
       "\n",
       "                                           sequences  \n",
       "0  [AGAAAGCTCCTCCAGTTCAC, TGATCGCTTTGACTGTGACA, A...  \n",
       "1  [GTAGAGCCTCCAGAACTGTG, AGGTTTATCCAGATGAACTG, C...  \n",
       "2  [CCAAGGCGTCCTCAGACCAG, AGCTCCAGGAAGGACCCCCG, T...  \n",
       "3  [GGCGTCAGTCGAGGAGTCAG, GCCAGCACTTCAGCTCACCG, G...  \n",
       "4  [ATAAGGCACTCACATCCACC, GCTTGTCCCTAACACTCAGA, G...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test building cell-guide matrix on subset of larger data frame\n",
    "phenodata_test_df = phenodata_df.head(5)\n",
    "phenodata_test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8ee48733-9f28-47d2-ab5e-427d26a5afdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_sequences = pd.Series(phenodata_test_df['sequences'].sum()).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6b262f75-b7fa-43bf-9a35-d34c7e8dd374",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2\n",
       "1    AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2\n",
       "2    AAACCTGCAAACAACA-1_1A_1_SI-GA-E2\n",
       "3    AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2\n",
       "4    AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2\n",
       "Name: cell, dtype: object"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_names = phenodata_test_df['cell']\n",
    "cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "86c1a4d2-89a7-4268-a6ff-0558ea864092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      1\n",
       "1      1\n",
       "2      1\n",
       "3      1\n",
       "4      1\n",
       "      ..\n",
       "225    0\n",
       "226    0\n",
       "227    0\n",
       "228    0\n",
       "229    0\n",
       "Length: 230, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.Series(guide_sequences).isin(phenodata_test_df['sequences'].loc[0]).astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f85584c6-0576-48a3-8ad4-4cadfaf10239",
   "metadata": {},
   "outputs": [],
   "source": [
    "def guides_present(cell_guide_list):\n",
    "    return pd.Series(guide_sequences).isin(cell_guide_list).astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c537dc4d-fef1-4b55-95d4-b8002a1facaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell</th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAAAGCTCCTCCAGTTCAC</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATCGCTTTGACTGTGACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAATAAAGAACAGAACACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTAAATTGAGACCTCAGGAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTTCCCCCCACCAATAACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGAGGAATCATCCTACCGGG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCAACTGGATGACACAG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTAGGCCTCCATATTCCTA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCGGGCCGAAGGTGCGAGG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGAGTACGCTCGGGAGCCCG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell                  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 1   \n",
       "TGATCGCTTTGACTGTGACA                                 1   \n",
       "ACAATAAAGAACAGAACACA                                 1   \n",
       "GTAAATTGAGACCTCAGGAG                                 1   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "...                                                ...   \n",
       "GGAGGAATCATCCTACCGGG                                 0   \n",
       "GCTGCAACTGGATGACACAG                                 0   \n",
       "GCTAGGCCTCCATATTCCTA                                 0   \n",
       "CCCGGGCCGAAGGTGCGAGG                                 0   \n",
       "GGAGTACGCTCGGGAGCCCG                                 0   \n",
       "\n",
       "cell                  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "...                                                ...   \n",
       "GGAGGAATCATCCTACCGGG                                 0   \n",
       "GCTGCAACTGGATGACACAG                                 0   \n",
       "GCTAGGCCTCCATATTCCTA                                 0   \n",
       "CCCGGGCCGAAGGTGCGAGG                                 0   \n",
       "GGAGTACGCTCGGGAGCCCG                                 0   \n",
       "\n",
       "cell                  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "...                                                ...   \n",
       "GGAGGAATCATCCTACCGGG                                 0   \n",
       "GCTGCAACTGGATGACACAG                                 0   \n",
       "GCTAGGCCTCCATATTCCTA                                 0   \n",
       "CCCGGGCCGAAGGTGCGAGG                                 0   \n",
       "GGAGTACGCTCGGGAGCCCG                                 0   \n",
       "\n",
       "cell                  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "...                                                ...   \n",
       "GGAGGAATCATCCTACCGGG                                 0   \n",
       "GCTGCAACTGGATGACACAG                                 0   \n",
       "GCTAGGCCTCCATATTCCTA                                 0   \n",
       "CCCGGGCCGAAGGTGCGAGG                                 0   \n",
       "GGAGTACGCTCGGGAGCCCG                                 0   \n",
       "\n",
       "cell                  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  \n",
       "TGATCGCTTTGACTGTGACA                                 0  \n",
       "ACAATAAAGAACAGAACACA                                 0  \n",
       "GTAAATTGAGACCTCAGGAG                                 0  \n",
       "TCTTCCCCCCACCAATAACA                                 0  \n",
       "...                                                ...  \n",
       "GGAGGAATCATCCTACCGGG                                 1  \n",
       "GCTGCAACTGGATGACACAG                                 1  \n",
       "GCTAGGCCTCCATATTCCTA                                 1  \n",
       "CCCGGGCCGAAGGTGCGAGG                                 1  \n",
       "GGAGTACGCTCGGGAGCCCG                                 1  \n",
       "\n",
       "[230 rows x 5 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_cell_guide_matrix = phenodata_test_df['sequences'].apply(guides_present)\n",
    "test_cell_guide_matrix.index = cell_names\n",
    "test_cell_guide_matrix.columns = guide_sequences\n",
    "test_cell_guide_matrix = test_cell_guide_matrix.T\n",
    "test_cell_guide_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "30e85c62-49f7-4d16-ba08-18182f71a6fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_cell_guide_matrix.to_hdf('./../data/test_cell_guide_matrix.h5', key='df', mode='w')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fede7c2c-1497-41a0-ba10-045e03424750",
   "metadata": {},
   "source": [
    "### Read in Cell-Guide Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8c184e67-090d-43c6-aea4-3f2dd8f774f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell</th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTATATGAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACAACGATGG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGATAATG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAAAGCTCCTCCAGTTCAC</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATCGCTTTGACTGTGACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAATAAAGAACAGAACACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTAAATTGAGACCTCAGGAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTTCCCCCCACCAATAACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  205797 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell                  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 1   \n",
       "TGATCGCTTTGACTGTGACA                                 1   \n",
       "ACAATAAAGAACAGAACACA                                 1   \n",
       "GTAAATTGAGACCTCAGGAG                                 1   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTATATGAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2  ...  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  ...   \n",
       "TGATCGCTTTGACTGTGACA                                 0  ...   \n",
       "ACAATAAAGAACAGAACACA                                 0  ...   \n",
       "GTAAATTGAGACCTCAGGAG                                 0  ...   \n",
       "TCTTCCCCCCACCAATAACA                                 0  ...   \n",
       "\n",
       "cell                  TTTGTCACAACGATGG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGATAATG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  \n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  \n",
       "TGATCGCTTTGACTGTGACA                                 0  \n",
       "ACAATAAAGAACAGAACACA                                 0  \n",
       "GTAAATTGAGACCTCAGGAG                                 0  \n",
       "TCTTCCCCCCACCAATAACA                                 0  \n",
       "\n",
       "[5 rows x 205797 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in cell-guide matrix\n",
    "cell_guide_matrix = pd.read_hdf('./../data/cell_guide_matrix.h5')\n",
    "cell_guide_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c0fa4bc-8992-4df4-ad5c-021ebea8ed15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13186, 205797)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get shape of cell guide matrix\n",
    "cell_guide_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6e220250-38ed-4e4e-97e7-add20cfe8442",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>[26606551, 26606569, 'chr1', 0.032]</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>[26606660, 26606641, 'chr1', 0.008]</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>[10530445, 10530427, 'chr11', 0.032]</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>[10530690, 10530708, 'chr11', 0.032]</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>[124195027, 124195008, 'chr8', 0.008]</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr8</td>\n",
       "      <td>124191287</td>\n",
       "      <td>124191288</td>\n",
       "      <td>FAM83A_TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence                         genomic_coords  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG    [26606551, 26606569, 'chr1', 0.032]   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG    [26606660, 26606641, 'chr1', 0.008]   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   [10530445, 10530427, 'chr11', 0.032]   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   [10530690, 10530708, 'chr11', 0.032]   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  [124195027, 124195008, 'chr8', 0.008]   \n",
       "\n",
       "         start          end  chrom  evalue gRNAgroup.chr gRNAgroup.start  \\\n",
       "0   26606551.0   26606569.0   chr1   0.032          chr1        26605667   \n",
       "1   26606660.0   26606641.0   chr1   0.008          chr1        26605667   \n",
       "2   10530445.0   10530427.0  chr11   0.032         chr11        10530735   \n",
       "3   10530690.0   10530708.0  chr11   0.032         chr11        10530735   \n",
       "4  124195027.0  124195008.0   chr8   0.008          chr8       124191287   \n",
       "\n",
       "  gRNAgroup.stop     gRNAgroup  \n",
       "0       26605668  SH3BGRL3_TSS  \n",
       "1       26605668  SH3BGRL3_TSS  \n",
       "2       10530736  MTRNR2L8_TSS  \n",
       "3       10530736  MTRNR2L8_TSS  \n",
       "4      124191288    FAM83A_TSS  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_sequence_coords = pd.read_csv('./../data/group_sequence_coords.csv')\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cf03400b-0fa9-48fb-8ddc-7f93b7ffe3ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr.targetgene</th>\n",
       "      <th>start.targetgene</th>\n",
       "      <th>stop.targetgene</th>\n",
       "      <th>ENSG.targetgene</th>\n",
       "      <th>targetgene_short_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28034777</td>\n",
       "      <td>28034778</td>\n",
       "      <td>ENSG00000150051</td>\n",
       "      <td>MKX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28287976</td>\n",
       "      <td>28287977</td>\n",
       "      <td>ENSG00000169126</td>\n",
       "      <td>ARMC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28571017</td>\n",
       "      <td>28571018</td>\n",
       "      <td>ENSG00000150054</td>\n",
       "      <td>MPP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28821422</td>\n",
       "      <td>28821423</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "      <td>WAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28966271</td>\n",
       "      <td>28966272</td>\n",
       "      <td>ENSG00000095739</td>\n",
       "      <td>BAMBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr.targetgene  start.targetgene  stop.targetgene  ENSG.targetgene  \\\n",
       "0          chr10          28034777         28034778  ENSG00000150051   \n",
       "1          chr10          28287976         28287977  ENSG00000169126   \n",
       "2          chr10          28571017         28571018  ENSG00000150054   \n",
       "3          chr10          28821422         28821423  ENSG00000095787   \n",
       "4          chr10          28966271         28966272  ENSG00000095739   \n",
       "\n",
       "  targetgene_short_name  \n",
       "0                   MKX  \n",
       "1                 ARMC4  \n",
       "2                  MPP7  \n",
       "3                   WAC  \n",
       "4                 BAMBI  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('./../data/genes_coords.csv')\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5803c95b-75c3-445c-9613-8fd7e01ab6e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_close_genes(guide_row):\n",
    "    '''find close genes given guide information'''\n",
    "    \n",
    "    # get guide information\n",
    "    guide_start = guide_row['start']\n",
    "    guide_end = guide_row['end']\n",
    "    guide_chrom = guide_row['chrom']\n",
    "    \n",
    "    # filter genes that match chromosome\n",
    "    filtered_genes = genes[genes['chr.targetgene'] == guide_chrom]\n",
    "    \n",
    "    # filter genes within range of 100kb\n",
    "    left = guide_start - 100000\n",
    "    right = guide_end + 100000\n",
    "    filtered_genes = filtered_genes[(filtered_genes['start.targetgene'] > left) & (filtered_genes['stop.targetgene'] < right)]\n",
    "    \n",
    "    return list(filtered_genes['ENSG.targetgene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1733ea07-87dd-49b2-b058-f973df0ef3fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence        start          end  chrom\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG   26606551.0   26606569.0   chr1\n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG   26606660.0   26606641.0   chr1\n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   10530445.0   10530427.0  chr11\n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   10530690.0   10530708.0  chr11\n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  124195027.0  124195008.0   chr8"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter necessary columns\n",
    "group_sequence_coords = group_sequence_coords[['gRNA_group', 'spacer_sequence', 'start', 'end', 'chrom']]\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bac9f63b-8fef-4ba2-94a8-0378228c8d98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[ENSG00000188782, ENSG00000130695, ENSG0000014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[ENSG00000188782, ENSG00000130695, ENSG0000014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>[ENSG00000133805, ENSG00000255823, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>[ENSG00000133805, ENSG00000255823, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[ENSG00000147689, ENSG00000189376, ENSG0000016...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence        start          end  chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG   26606551.0   26606569.0   chr1   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG   26606660.0   26606641.0   chr1   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   10530445.0   10530427.0  chr11   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   10530690.0   10530708.0  chr11   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  124195027.0  124195008.0   chr8   \n",
       "\n",
       "                                      proximal_genes  \n",
       "0  [ENSG00000188782, ENSG00000130695, ENSG0000014...  \n",
       "1  [ENSG00000188782, ENSG00000130695, ENSG0000014...  \n",
       "2  [ENSG00000133805, ENSG00000255823, ENSG0000011...  \n",
       "3  [ENSG00000133805, ENSG00000255823, ENSG0000011...  \n",
       "4  [ENSG00000147689, ENSG00000189376, ENSG0000016...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get proximal genes for each guide sequence\n",
    "group_sequence_coords['proximal_genes'] = group_sequence_coords.apply(find_close_genes, axis=1)\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2d223639-8713-4cfa-a203-3b938598f1b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000188782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000130695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000142669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000158062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence       start         end chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "1  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "2  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "3  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "4  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "\n",
       "    proximal_genes  \n",
       "0  ENSG00000188782  \n",
       "1  ENSG00000130695  \n",
       "2  ENSG00000142669  \n",
       "3  ENSG00000158062  \n",
       "4  ENSG00000169442  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explode dataframe to get guide-gene pairs\n",
    "guide_gene_pairs = group_sequence_coords.explode('proximal_genes', ignore_index=True)\n",
    "guide_gene_pairs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "69a625a7-11e2-4bfe-9b95-60f312986c30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34817, 6)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get shape of guide gene pairs\n",
    "guide_gene_pairs.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce59161-0f75-44c4-abca-7377ca273107",
   "metadata": {},
   "source": [
    "### Baseline GLM (Generalized Linear Model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "267c45b5-6dd8-4ab3-afea-0ca498f5339f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gRNA_group            0\n",
       "spacer_sequence       0\n",
       "start                12\n",
       "end                  12\n",
       "chrom                12\n",
       "proximal_genes     3021\n",
       "dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_gene_pairs.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1a0beafb-e234-42eb-8f33-33acbf479ff4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000188782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000130695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000142669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000158062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence       start         end chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "1  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "2  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "3  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "4  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "\n",
       "    proximal_genes  \n",
       "0  ENSG00000188782  \n",
       "1  ENSG00000130695  \n",
       "2  ENSG00000142669  \n",
       "3  ENSG00000158062  \n",
       "4  ENSG00000169442  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_gene_pairs = guide_gene_pairs.dropna()\n",
    "guide_gene_pairs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "310d100f-906a-48cc-b090-96ab1f607247",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31796, 6)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_gene_pairs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "88e45e81-3f26-4ada-8c0d-0f57a0999173",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2    0\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2    0\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2    0\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2    0\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2    0\n",
       "                                   ..\n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9    0\n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9    0\n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    0\n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9    1\n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9    0\n",
       "Name: ENSG00000130695, Length: 207324, dtype: int64"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_vector = cell_guide_matrix.loc[guide_gene_pairs['spacer_sequence'][0]]\n",
    "output_vector = count_data.loc[guide_gene_pairs['proximal_genes'][1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8629bacb-188e-4458-8346-f3fe1d9a362d",
   "metadata": {},
   "source": [
    "### Determine Guide Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0caad155-b9c9-48a9-8667-ee9ffe45ec46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    AGAAAGCTCCTCCAGTTCAC\n",
       "1    TGATCGCTTTGACTGTGACA\n",
       "2    ACAATAAAGAACAGAACACA\n",
       "3    GTAAATTGAGACCTCAGGAG\n",
       "4    TCTTCCCCCCACCAATAACA\n",
       "dtype: object"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get guide sequences\n",
    "guide_sequences = pd.Series(cell_guide_matrix.index)\n",
    "guide_sequences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d2bfc866-6c53-4dfd-a4a8-a3e5be11113b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>[26606551, 26606569, 'chr1', 0.032]</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>[26606660, 26606641, 'chr1', 0.008]</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>[10530445, 10530427, 'chr11', 0.032]</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>[10530690, 10530708, 'chr11', 0.032]</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>[124195027, 124195008, 'chr8', 0.008]</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr8</td>\n",
       "      <td>124191287</td>\n",
       "      <td>124191288</td>\n",
       "      <td>FAM83A_TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence                         genomic_coords  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG    [26606551, 26606569, 'chr1', 0.032]   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG    [26606660, 26606641, 'chr1', 0.008]   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   [10530445, 10530427, 'chr11', 0.032]   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   [10530690, 10530708, 'chr11', 0.032]   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  [124195027, 124195008, 'chr8', 0.008]   \n",
       "\n",
       "         start          end  chrom  evalue gRNAgroup.chr gRNAgroup.start  \\\n",
       "0   26606551.0   26606569.0   chr1   0.032          chr1        26605667   \n",
       "1   26606660.0   26606641.0   chr1   0.008          chr1        26605667   \n",
       "2   10530445.0   10530427.0  chr11   0.032         chr11        10530735   \n",
       "3   10530690.0   10530708.0  chr11   0.032         chr11        10530735   \n",
       "4  124195027.0  124195008.0   chr8   0.008          chr8       124191287   \n",
       "\n",
       "  gRNAgroup.stop     gRNAgroup  \n",
       "0       26605668  SH3BGRL3_TSS  \n",
       "1       26605668  SH3BGRL3_TSS  \n",
       "2       10530736  MTRNR2L8_TSS  \n",
       "3       10530736  MTRNR2L8_TSS  \n",
       "4      124191288    FAM83A_TSS  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in group sequence coordinates file\n",
    "group_sequence_coords = pd.read_csv('./../data/group_sequence_coords.csv')\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f43e1530-1143-44ff-82bf-bb2c27f03b5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1963    244486508.0\n",
       "Name: start, dtype: float64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# match guide sequences to genomic coordinates\n",
    "(group_sequence_coords[guide_sequences.loc[5] == group_sequence_coords['spacer_sequence']]['start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7ef58e39-1c96-4696-acc5-ba8f756038e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_genomic_coordinates(spacer_sequence):\n",
    "    '''helper function to match spacer sequence to corresponding genomic coordinates'''\n",
    "    \n",
    "    guide_row = group_sequence_coords[group_sequence_coords['spacer_sequence'] == spacer_sequence]\n",
    "    return guide_row['start'], guide_row['end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "41f1ac30-8fa4-47e0-9c1c-2204a71cd0b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence        start          end  chrom\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG   26606551.0   26606569.0   chr1\n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG   26606660.0   26606641.0   chr1\n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   10530445.0   10530427.0  chr11\n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   10530690.0   10530708.0  chr11\n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  124195027.0  124195008.0   chr8"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_sequence_coords = group_sequence_coords[['gRNA_group', 'spacer_sequence', 'start', 'end', 'chrom']]\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b2fc5320-fc50-4a4d-9cb0-efc5b68b0554",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGATCGCTTTGACTGTGACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAATAAAGAACAGAACACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTAAATTGAGACCTCAGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCTTCCCCCCACCAATAACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        spacer_sequence\n",
       "0  AGAAAGCTCCTCCAGTTCAC\n",
       "1  TGATCGCTTTGACTGTGACA\n",
       "2  ACAATAAAGAACAGAACACA\n",
       "3  GTAAATTGAGACCTCAGGAG\n",
       "4  TCTTCCCCCCACCAATAACA"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create dataframe of guide sequences\n",
    "guide_sequences_df = pd.DataFrame(guide_sequences, columns=['spacer_sequence'])\n",
    "guide_sequences_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f745b950-3473-4d41-90d4-1d70e55e9037",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017.0</td>\n",
       "      <td>23345036.0</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGATCGCTTTGACTGTGACA</td>\n",
       "      <td>chr1.11183_top_two</td>\n",
       "      <td>205720521.0</td>\n",
       "      <td>205720502.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAATAAAGAACAGAACACA</td>\n",
       "      <td>chr1.11293_top_two</td>\n",
       "      <td>207020873.0</td>\n",
       "      <td>207020892.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTAAATTGAGACCTCAGGAG</td>\n",
       "      <td>chr11.1791_top_two</td>\n",
       "      <td>34847978.0</td>\n",
       "      <td>34847959.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCTTCCCCCCACCAATAACA</td>\n",
       "      <td>chr1.12598_top_two</td>\n",
       "      <td>235029683.0</td>\n",
       "      <td>235029664.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        spacer_sequence          gRNA_group        start          end  chrom\n",
       "0  AGAAAGCTCCTCCAGTTCAC   chr10.845_top_two   23345017.0   23345036.0  chr10\n",
       "1  TGATCGCTTTGACTGTGACA  chr1.11183_top_two  205720521.0  205720502.0   chr1\n",
       "2  ACAATAAAGAACAGAACACA  chr1.11293_top_two  207020873.0  207020892.0   chr1\n",
       "3  GTAAATTGAGACCTCAGGAG  chr11.1791_top_two   34847978.0   34847959.0  chr11\n",
       "4  TCTTCCCCCCACCAATAACA  chr1.12598_top_two  235029683.0  235029664.0   chr1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge guide sequences with group sequence coordinates\n",
    "merged_df = guide_sequences_df.merge(group_sequence_coords, how='left')\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "75755796-5356-4200-a3dc-4e8b5b188bd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13186, 5)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check dataframe shape\n",
    "merged_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b733701d-3244-4e3d-9eb4-cc1439c11599",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "spacer_sequence     0\n",
       "gRNA_group         26\n",
       "start              38\n",
       "end                38\n",
       "chrom              38\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# analyze missingness in dataset\n",
    "merged_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "00090a95-2f77-4fdd-ac59-7829f1a89b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop nan values (cannot be used for analysis)\n",
    "merged_df = merged_df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8bd91228-79f2-4ac2-a6ca-35cbd2786ab6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13148, 5)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check new shape of dataset\n",
    "merged_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e1e2ce5d-ee84-4e85-992d-96fad7bcfa42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGATCGCTTTGACTGTGACA</td>\n",
       "      <td>chr1.11183_top_two</td>\n",
       "      <td>205720521</td>\n",
       "      <td>205720502</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAATAAAGAACAGAACACA</td>\n",
       "      <td>chr1.11293_top_two</td>\n",
       "      <td>207020873</td>\n",
       "      <td>207020892</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTAAATTGAGACCTCAGGAG</td>\n",
       "      <td>chr11.1791_top_two</td>\n",
       "      <td>34847978</td>\n",
       "      <td>34847959</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCTTCCCCCCACCAATAACA</td>\n",
       "      <td>chr1.12598_top_two</td>\n",
       "      <td>235029683</td>\n",
       "      <td>235029664</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        spacer_sequence          gRNA_group      start        end  chrom\n",
       "0  AGAAAGCTCCTCCAGTTCAC   chr10.845_top_two   23345017   23345036  chr10\n",
       "1  TGATCGCTTTGACTGTGACA  chr1.11183_top_two  205720521  205720502   chr1\n",
       "2  ACAATAAAGAACAGAACACA  chr1.11293_top_two  207020873  207020892   chr1\n",
       "3  GTAAATTGAGACCTCAGGAG  chr11.1791_top_two   34847978   34847959  chr11\n",
       "4  TCTTCCCCCCACCAATAACA  chr1.12598_top_two  235029683  235029664   chr1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert datatypes for merged dataframe\n",
    "merged_df['start'] = merged_df['start'].astype(np.int64)\n",
    "merged_df['end'] = merged_df['end'].astype(np.int64)\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f0a1d2d6-8c68-411e-bfb3-279696d11431",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAAGTTTGTTGAAAGTGGCG',\n",
       " 'CCTCAGTGCATAAGGCCAAT',\n",
       " 'GTTTGCTGAAGATAAATGCA',\n",
       " 'AACACAAGCAATGCGACAAC',\n",
       " 'AGACCGCCGTGAAGACCCCG',\n",
       " 'GCTTCTGTGTGACTCATACG',\n",
       " 'GTAGAGCTCAGATAAGAGCT',\n",
       " 'GAAAGCTCCTCCAGTTCACG',\n",
       " 'CATGTTGGAGATGTGCCTTA',\n",
       " 'CTGCCAAGCTCAATTGTCCT',\n",
       " 'GTCGCCATCTTGTTTATGTT',\n",
       " 'GCCCGGAGGAGAAGCGTGAG',\n",
       " 'GGTCACGGCGCGCTCACAAG',\n",
       " 'AGAAAAAGCAATCTTGACTG',\n",
       " 'TAGGTGCTGGCTGAAAAACG',\n",
       " 'GTAGGAGTCCCTCCAGGAGA',\n",
       " 'GTAAATCACGTGGCCATGCC',\n",
       " 'TCTATTGGTCTGTTAGGTGC',\n",
       " 'GACAAAGTCACTCTGTCCCA',\n",
       " 'ATATGATCGTTTGGTCACAA',\n",
       " 'GTGACGCGTGAAATGTAACG',\n",
       " 'GAGACAGCTAAGTGACCAAA',\n",
       " 'CATGGCCACGTGATTTACTC',\n",
       " 'GGCAGACTATTGCTGCTGGA',\n",
       " 'TGGAGGGATTCTGTACAGAG',\n",
       " 'AGGACAATTGAGCTTGGCAG',\n",
       " 'TTGTTTATGTTGGGTTTGGC',\n",
       " 'ATAGGCGGCTGACCACCCAG',\n",
       " 'TACATCTGTTCACATTTCAC',\n",
       " 'TGAACAGATGTATACACTTC',\n",
       " 'CCATAGACTGGGTATTTCAT',\n",
       " 'AAAAAGACGGAGGAAGTGCA',\n",
       " 'ACTTAAGGTAGTAAACTGAT']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test to find all guides within 1MB for first start site\n",
    "test_chrom = merged_df.loc[0]['chrom']\n",
    "test_filter = merged_df[merged_df['chrom'] == 'chr10']\n",
    "test_start = merged_df.loc[0]['start']\n",
    "test_filter = test_filter[test_filter['start'] != test_start]\n",
    "test_diffs = np.abs(test_start - test_filter['start'])\n",
    "test_matches = test_diffs <= 1000000\n",
    "list(test_filter[test_matches]['spacer_sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "07dcac41-c283-4307-8555-301f4bb9f21a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_proximal_guides(guide_row):\n",
    "    '''function to return list of proximal guides given guide information'''\n",
    "    \n",
    "    # get guide chromsome and start information \n",
    "    guide_start = guide_row['start']\n",
    "    guide_chrom = guide_row['chrom']\n",
    "    \n",
    "    # filter dataframe for chromosome and remove current guide\n",
    "    filtered_guides = merged_df[merged_df['chrom'] == guide_chrom]\n",
    "    filtered_guides = filtered_guides[filtered_guides['start'] != guide_start]\n",
    "    \n",
    "    # get 1MB proximal guide matches and return list of proximal guides\n",
    "    bp_diffs = np.abs(guide_start - filtered_guides['start'])\n",
    "    bp_matches = bp_diffs <= 1000000\n",
    "    return list(filtered_guides[bp_matches]['spacer_sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "394ffeb8-704a-4495-a491-f2ec4add8896",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_guides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[AAAGTTTGTTGAAAGTGGCG, CCTCAGTGCATAAGGCCAAT, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGATCGCTTTGACTGTGACA</td>\n",
       "      <td>chr1.11183_top_two</td>\n",
       "      <td>205720521</td>\n",
       "      <td>205720502</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[CTAGGTGGTGTGGGGAGTGA, TCTTGACTAACAGGACCCCA, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAATAAAGAACAGAACACA</td>\n",
       "      <td>chr1.11293_top_two</td>\n",
       "      <td>207020873</td>\n",
       "      <td>207020892</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[GTTGTCCTTCAGGGACAGTG, GCCAGCACTTCAGCTCACCG, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTAAATTGAGACCTCAGGAG</td>\n",
       "      <td>chr11.1791_top_two</td>\n",
       "      <td>34847978</td>\n",
       "      <td>34847959</td>\n",
       "      <td>chr11</td>\n",
       "      <td>[CCCGGGCCGAAGGTGCGAGG, TTACTGGGAGGGCATCAGAG, G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCTTCCCCCCACCAATAACA</td>\n",
       "      <td>chr1.12598_top_two</td>\n",
       "      <td>235029683</td>\n",
       "      <td>235029664</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[GAGAACCTATTCCTGTCCAG, GAAGGAAGTGGGCTTCCACA, T...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        spacer_sequence          gRNA_group      start        end  chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC   chr10.845_top_two   23345017   23345036  chr10   \n",
       "1  TGATCGCTTTGACTGTGACA  chr1.11183_top_two  205720521  205720502   chr1   \n",
       "2  ACAATAAAGAACAGAACACA  chr1.11293_top_two  207020873  207020892   chr1   \n",
       "3  GTAAATTGAGACCTCAGGAG  chr11.1791_top_two   34847978   34847959  chr11   \n",
       "4  TCTTCCCCCCACCAATAACA  chr1.12598_top_two  235029683  235029664   chr1   \n",
       "\n",
       "                                     proximal_guides  \n",
       "0  [AAAGTTTGTTGAAAGTGGCG, CCTCAGTGCATAAGGCCAAT, G...  \n",
       "1  [CTAGGTGGTGTGGGGAGTGA, TCTTGACTAACAGGACCCCA, G...  \n",
       "2  [GTTGTCCTTCAGGGACAGTG, GCCAGCACTTCAGCTCACCG, G...  \n",
       "3  [CCCGGGCCGAAGGTGCGAGG, TTACTGGGAGGGCATCAGAG, G...  \n",
       "4  [GAGAACCTATTCCTGTCCAG, GAAGGAAGTGGGCTTCCACA, T...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get proximal guides (within 1MB) for every guide in the data frame\n",
    "merged_df['proximal_guides'] = merged_df.apply(find_proximal_guides, axis=1)\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e967103-6d74-45c7-a750-14c264364425",
   "metadata": {},
   "source": [
    "### Determine Gene-Guide Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bf87c09f-6a11-4e11-82d5-4c7cda5ce3fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "spacer_sequence                                 AGAAAGCTCCTCCAGTTCAC\n",
       "gRNA_group                                         chr10.845_top_two\n",
       "start                                                       23345017\n",
       "end                                                         23345036\n",
       "chrom                                                          chr10\n",
       "proximal_guides    [AAAGTTTGTTGAAAGTGGCG, CCTCAGTGCATAAGGCCAAT, G...\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_guide = merged_df.loc[0]\n",
    "test_guide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1cd80bfa-d25d-4351-968f-b538d81f00c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     AAAGTTTGTTGAAAGTGGCG\n",
       "1     CCTCAGTGCATAAGGCCAAT\n",
       "2     GTTTGCTGAAGATAAATGCA\n",
       "3     AACACAAGCAATGCGACAAC\n",
       "4     AGACCGCCGTGAAGACCCCG\n",
       "5     GCTTCTGTGTGACTCATACG\n",
       "6     GTAGAGCTCAGATAAGAGCT\n",
       "7     GAAAGCTCCTCCAGTTCACG\n",
       "8     CATGTTGGAGATGTGCCTTA\n",
       "9     CTGCCAAGCTCAATTGTCCT\n",
       "10    GTCGCCATCTTGTTTATGTT\n",
       "11    GCCCGGAGGAGAAGCGTGAG\n",
       "12    GGTCACGGCGCGCTCACAAG\n",
       "13    AGAAAAAGCAATCTTGACTG\n",
       "14    TAGGTGCTGGCTGAAAAACG\n",
       "15    GTAGGAGTCCCTCCAGGAGA\n",
       "16    GTAAATCACGTGGCCATGCC\n",
       "17    TCTATTGGTCTGTTAGGTGC\n",
       "18    GACAAAGTCACTCTGTCCCA\n",
       "19    ATATGATCGTTTGGTCACAA\n",
       "20    GTGACGCGTGAAATGTAACG\n",
       "21    GAGACAGCTAAGTGACCAAA\n",
       "22    CATGGCCACGTGATTTACTC\n",
       "23    GGCAGACTATTGCTGCTGGA\n",
       "24    TGGAGGGATTCTGTACAGAG\n",
       "25    AGGACAATTGAGCTTGGCAG\n",
       "26    TTGTTTATGTTGGGTTTGGC\n",
       "27    ATAGGCGGCTGACCACCCAG\n",
       "28    TACATCTGTTCACATTTCAC\n",
       "29    TGAACAGATGTATACACTTC\n",
       "30    CCATAGACTGGGTATTTCAT\n",
       "31    AAAAAGACGGAGGAAGTGCA\n",
       "32    ACTTAAGGTAGTAAACTGAT\n",
       "dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_guide_start = test_guide['start']\n",
    "pd.Series(test_guide['proximal_guides'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61560e71-3712-4a8c-a58c-c1dadbedc688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_guides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AACACAAGCAATGCGACAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        spacer_sequence         gRNA_group     start       end  chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036  chr10   \n",
       "1  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036  chr10   \n",
       "2  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036  chr10   \n",
       "3  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036  chr10   \n",
       "4  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036  chr10   \n",
       "\n",
       "        proximal_guides  \n",
       "0  AAAGTTTGTTGAAAGTGGCG  \n",
       "1  CCTCAGTGCATAAGGCCAAT  \n",
       "2  GTTTGCTGAAGATAAATGCA  \n",
       "3  AACACAAGCAATGCGACAAC  \n",
       "4  AGACCGCCGTGAAGACCCCG  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explode to get guide pairs in individual rows\n",
    "exploded_df = merged_df.explode('proximal_guides', ignore_index=True)\n",
    "exploded_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5d022876-ead9-4e40-9605-ad6f4ba3eb53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(291647, 6)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exploded_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7d814889-d057-42f3-b958-bc3ea242a0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spacer_sequence_x</th>\n",
       "      <th>gRNA_group_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>proximal_guides</th>\n",
       "      <th>spacer_sequence_y</th>\n",
       "      <th>gRNA_group_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>chrom_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      spacer_sequence_x          gRNA_group_x   start_x     end_x chrom_x  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036   chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940   chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624   chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121   chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219   chr10   \n",
       "\n",
       "        proximal_guides     spacer_sequence_y       gRNA_group_y   start_y  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685   \n",
       "\n",
       "      end_y chrom_y  \n",
       "0  22519666   chr10  \n",
       "1  22519666   chr10  \n",
       "2  22519666   chr10  \n",
       "3  22519666   chr10  \n",
       "4  22519666   chr10  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get start and end coordinates for proximal guides\n",
    "join_df = merged_df.drop(['proximal_guides'], axis=1)\n",
    "guide_pairs_df = exploded_df.merge(join_df, left_on='proximal_guides', right_on='spacer_sequence')\n",
    "guide_pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "59c1c283-02f3-4770-94a7-09b46b973249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10  \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10  \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10  \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10  \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rename columns for guide pairs\n",
    "colnames = ['g1_spacer_sequence', 'g1_gRNA_group', 'g1_start', 'g1_end', 'g1_chrom', 'g1_proximal',\n",
    "            'g2_spacer_sequence', 'g2_gRNA_group', 'g2_start', 'g2_end', 'g2_chrom']\n",
    "guide_pairs_df.columns = colnames\n",
    "guide_pairs_df = guide_pairs_df.drop(['g1_proximal'], axis=1)\n",
    "guide_pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6ba75e58-0273-4d8d-b4a1-bec01007b7ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>22519685</td>\n",
       "      <td>23345036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22519666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22519666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>22519685</td>\n",
       "      <td>23144121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22519666</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right  \n",
       "0  22519685  23345036  \n",
       "1  22422921  22519666  \n",
       "2  22306605  22519666  \n",
       "3  22519685  23144121  \n",
       "4  22518200  22519666  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get leftmost and rightmost coordinates for each guide pair\n",
    "guide_pairs_df['left'] = guide_pairs_df[['g1_start', 'g2_start']].min(axis=1)\n",
    "guide_pairs_df['right'] = guide_pairs_df[['g1_end', 'g2_end']].max(axis=1)\n",
    "guide_pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b0d9a1e9-5aae-4cf7-9602-032b63925a1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21422921</td>\n",
       "      <td>23519666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21306605</td>\n",
       "      <td>23519666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24144121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21518200</td>\n",
       "      <td>23519666</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right  \n",
       "0  21519685  24345036  \n",
       "1  21422921  23519666  \n",
       "2  21306605  23519666  \n",
       "3  21519685  24144121  \n",
       "4  21518200  23519666  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add 1 megabase region for finding genes\n",
    "guide_pairs_df['left'] = guide_pairs_df['left'] - 1000000\n",
    "guide_pairs_df['right'] = guide_pairs_df['right'] + 1000000\n",
    "guide_pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "292ea5dd-6869-49fc-bde5-80e1a9c9f8e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "      <th>general_group</th>\n",
       "      <th>chr.targetgene</th>\n",
       "      <th>start.targetgene</th>\n",
       "      <th>stop.targetgene</th>\n",
       "      <th>ENSG.targetgene</th>\n",
       "      <th>targetgene_short_name</th>\n",
       "      <th>strand.targetgene</th>\n",
       "      <th>pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28034777</td>\n",
       "      <td>28034778</td>\n",
       "      <td>ENSG00000150051</td>\n",
       "      <td>MKX</td>\n",
       "      <td>-</td>\n",
       "      <td>MKX:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28287976</td>\n",
       "      <td>28287977</td>\n",
       "      <td>ENSG00000169126</td>\n",
       "      <td>ARMC4</td>\n",
       "      <td>-</td>\n",
       "      <td>ARMC4:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28571017</td>\n",
       "      <td>28571018</td>\n",
       "      <td>ENSG00000150054</td>\n",
       "      <td>MPP7</td>\n",
       "      <td>-</td>\n",
       "      <td>MPP7:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28821422</td>\n",
       "      <td>28821423</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "      <td>WAC</td>\n",
       "      <td>+</td>\n",
       "      <td>WAC:bassik_mch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>NTC</td>\n",
       "      <td>bassik_mch</td>\n",
       "      <td>NTC</td>\n",
       "      <td>chr10</td>\n",
       "      <td>28966271</td>\n",
       "      <td>28966272</td>\n",
       "      <td>ENSG00000095739</td>\n",
       "      <td>BAMBI</td>\n",
       "      <td>+</td>\n",
       "      <td>BAMBI:bassik_mch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gRNAgroup.chr gRNAgroup.start gRNAgroup.stop   gRNAgroup general_group  \\\n",
       "0           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "1           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "2           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "3           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "4           NTC             NTC            NTC  bassik_mch           NTC   \n",
       "\n",
       "  chr.targetgene  start.targetgene  stop.targetgene  ENSG.targetgene  \\\n",
       "0          chr10          28034777         28034778  ENSG00000150051   \n",
       "1          chr10          28287976         28287977  ENSG00000169126   \n",
       "2          chr10          28571017         28571018  ENSG00000150054   \n",
       "3          chr10          28821422         28821423  ENSG00000095787   \n",
       "4          chr10          28966271         28966272  ENSG00000095739   \n",
       "\n",
       "  targetgene_short_name strand.targetgene             pairs  \n",
       "0                   MKX                 -    MKX:bassik_mch  \n",
       "1                 ARMC4                 -  ARMC4:bassik_mch  \n",
       "2                  MPP7                 -   MPP7:bassik_mch  \n",
       "3                   WAC                 +    WAC:bassik_mch  \n",
       "4                 BAMBI                 +  BAMBI:bassik_mch  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in dataframe of intended target regions\n",
    "dtypes = {\n",
    "    'gRNAgroup.start': str,\n",
    "    'gRNAgroup.stop': str\n",
    "}\n",
    "\n",
    "intended_targets_df = pd.read_csv('/iblm/netapp/data1/external/Gasperini2019/suppl/' + \\\n",
    "                                  'GSE120861_gene_gRNAgroup_pair_table.at_scale.txt', sep='\\t', dtype=dtypes)\n",
    "intended_targets_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "246fa4fe-ceda-4c5e-8349-01285188127c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr.targetgene</th>\n",
       "      <th>start.targetgene</th>\n",
       "      <th>stop.targetgene</th>\n",
       "      <th>ENSG.targetgene</th>\n",
       "      <th>targetgene_short_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28034777</td>\n",
       "      <td>28034778</td>\n",
       "      <td>ENSG00000150051</td>\n",
       "      <td>MKX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28287976</td>\n",
       "      <td>28287977</td>\n",
       "      <td>ENSG00000169126</td>\n",
       "      <td>ARMC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28571017</td>\n",
       "      <td>28571018</td>\n",
       "      <td>ENSG00000150054</td>\n",
       "      <td>MPP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28821422</td>\n",
       "      <td>28821423</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "      <td>WAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28966271</td>\n",
       "      <td>28966272</td>\n",
       "      <td>ENSG00000095739</td>\n",
       "      <td>BAMBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr.targetgene  start.targetgene  stop.targetgene  ENSG.targetgene  \\\n",
       "0          chr10          28034777         28034778  ENSG00000150051   \n",
       "1          chr10          28287976         28287977  ENSG00000169126   \n",
       "2          chr10          28571017         28571018  ENSG00000150054   \n",
       "3          chr10          28821422         28821423  ENSG00000095787   \n",
       "4          chr10          28966271         28966272  ENSG00000095739   \n",
       "\n",
       "  targetgene_short_name  \n",
       "0                   MKX  \n",
       "1                 ARMC4  \n",
       "2                  MPP7  \n",
       "3                   WAC  \n",
       "4                 BAMBI  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = intended_targets_df[['chr.targetgene', 'start.targetgene', 'stop.targetgene', \n",
    "                               'ENSG.targetgene', 'targetgene_short_name']]\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "794f9668-9e0c-42a0-9cce-8ad479d72e72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1086564, 5)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get shape of gene dataframe\n",
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a7811cc8-6e84-48a0-a4e6-bfc209fd9fef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18389, 5)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicates and get new dataframe shape \n",
    "gene_df = gene_df.drop_duplicates()\n",
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "507b8450-259e-4973-993e-4e32ca8f7623",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g1_spacer_sequence    AGAAAGCTCCTCCAGTTCAC\n",
       "g1_gRNA_group            chr10.845_top_two\n",
       "g1_start                          23345017\n",
       "g1_end                            23345036\n",
       "g1_chrom                             chr10\n",
       "g2_spacer_sequence    AAAGTTTGTTGAAAGTGGCG\n",
       "g2_gRNA_group            chr10.791_top_two\n",
       "g2_start                          22519685\n",
       "g2_end                            22519666\n",
       "g2_chrom                             chr10\n",
       "left                              21519685\n",
       "right                             24345036\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data = guide_pairs_df.loc[0]\n",
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4b9ab897-041a-4ffc-8372-89b756ab6014",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_matches = gene_df[(gene_df['start.targetgene'] > test_data['left']) & (gene_df['stop.targetgene'] < test_data['right'])]\n",
    "test_gene_names = list(test_matches['ENSG.targetgene'])\n",
    "test_short_gene_names = list(test_matches['targetgene_short_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "068edf5d-7598-4a87-aae2-326738f4f9e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write current dataframes to CSV files\n",
    "gene_df.to_csv('./../data/genes_coords.csv', index=False)\n",
    "guide_pairs_df.to_csv('./../data/guide_pairs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3f549830-eac3-4f72-bd4d-fa2f32ff0af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_proximal_genes(guide_pair_row):\n",
    "    '''function to get proximal genes and short gene names given guide pair information'''\n",
    "    \n",
    "    gene_matches = gene_df[(gene_df['start.targetgene'] > guide_pair_row['left']) & \n",
    "                           (gene_df['stop.targetgene'] < guide_pair_row['right'])]\n",
    "    gene_names = list(gene_matches['ENSG.targetgene'])\n",
    "    short_gene_names = list(gene_matches['targetgene_short_name'])\n",
    "    return pd.Series([gene_names, short_gene_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ceb703c6-f1b4-40d2-acb2-b2a8a655b62e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>[ENSG00000204682, ENSG00000180592, ENSG0000007...</td>\n",
       "      <td>[CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21422921</td>\n",
       "      <td>23519666</td>\n",
       "      <td>[ENSG00000204683, ENSG00000204682, ENSG0000018...</td>\n",
       "      <td>[C10orf113, CASC10, SKIDA1, MLLT10, DNAJC1, EB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21306605</td>\n",
       "      <td>23519666</td>\n",
       "      <td>[ENSG00000204683, ENSG00000204682, ENSG0000018...</td>\n",
       "      <td>[C10orf113, CASC10, SKIDA1, MLLT10, DNAJC1, EB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24144121</td>\n",
       "      <td>[ENSG00000204682, ENSG00000180592, ENSG0000007...</td>\n",
       "      <td>[CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21518200</td>\n",
       "      <td>23519666</td>\n",
       "      <td>[ENSG00000204682, ENSG00000180592, ENSG0000007...</td>\n",
       "      <td>[CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right                                                  0  \\\n",
       "0  21519685  24345036  [ENSG00000204682, ENSG00000180592, ENSG0000007...   \n",
       "1  21422921  23519666  [ENSG00000204683, ENSG00000204682, ENSG0000018...   \n",
       "2  21306605  23519666  [ENSG00000204683, ENSG00000204682, ENSG0000018...   \n",
       "3  21519685  24144121  [ENSG00000204682, ENSG00000180592, ENSG0000007...   \n",
       "4  21518200  23519666  [ENSG00000204682, ENSG00000180592, ENSG0000007...   \n",
       "\n",
       "                                                   1  \n",
       "0  [CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...  \n",
       "1  [C10orf113, CASC10, SKIDA1, MLLT10, DNAJC1, EB...  \n",
       "2  [C10orf113, CASC10, SKIDA1, MLLT10, DNAJC1, EB...  \n",
       "3  [CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...  \n",
       "4  [CASC10, SKIDA1, MLLT10, DNAJC1, EBLN1, COMMD3...  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_output = guide_pairs_df.head().apply(get_proximal_genes, axis=1)\n",
    "pd.concat([guide_pairs_df.head(), test_output], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4ecb7504-424e-4111-bd62-13a1046bcab4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(291584, 12)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "39073d5d-c8b6-4be2-88e4-580a67d1e78c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18389, 5)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b5f5885a-9bd0-49d4-bee3-2c25c22b98e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>proximal_gene_names</th>\n",
       "      <th>proximal_short_gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "      <td>['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21422921</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204683', 'ENSG00000204682', 'ENSG00...</td>\n",
       "      <td>['C10orf113', 'CASC10', 'SKIDA1', 'MLLT10', 'D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21306605</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204683', 'ENSG00000204682', 'ENSG00...</td>\n",
       "      <td>['C10orf113', 'CASC10', 'SKIDA1', 'MLLT10', 'D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24144121</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "      <td>['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21518200</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "      <td>['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right                                proximal_gene_names  \\\n",
       "0  21519685  24345036  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...   \n",
       "1  21422921  23519666  ['ENSG00000204683', 'ENSG00000204682', 'ENSG00...   \n",
       "2  21306605  23519666  ['ENSG00000204683', 'ENSG00000204682', 'ENSG00...   \n",
       "3  21519685  24144121  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...   \n",
       "4  21518200  23519666  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...   \n",
       "\n",
       "                           proximal_short_gene_names  \n",
       "0  ['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...  \n",
       "1  ['C10orf113', 'CASC10', 'SKIDA1', 'MLLT10', 'D...  \n",
       "2  ['C10orf113', 'CASC10', 'SKIDA1', 'MLLT10', 'D...  \n",
       "3  ['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...  \n",
       "4  ['CASC10', 'SKIDA1', 'MLLT10', 'DNAJC1', 'EBLN...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_genes = pd.read_csv('./../data/guide_pairs_proximal_genes.csv')\n",
    "guide_pairs_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65b79659-1d69-4f1a-bcbe-2b08fbc28def",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(291584, 14)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get shape of dataframe\n",
    "guide_pairs_genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1186984-b162-42cd-b418-a97bec415109",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_pairs_genes = guide_pairs_genes.drop(['proximal_short_gene_names'], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "93f3cf72-8d3f-4a10-b106-58992d2e36c5",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'strip'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-1c28b4642a03>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proximal_gene_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proximal_gene_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m                                            \u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'[]'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m', '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/jupyter/lib/python3.9/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   4136\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4137\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4138\u001b[0;31m                 \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4139\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4140\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-1c28b4642a03>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proximal_gene_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'proximal_gene_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m                                            \u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'[]'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m', '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mguide_pairs_genes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'strip'"
     ]
    }
   ],
   "source": [
    "guide_pairs_genes['proximal_gene_names'] = guide_pairs_genes['proximal_gene_names']\\\n",
    "                                           .apply(lambda x: x.strip('[]').split(', '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee06d4f2-15b2-495c-b374-526561e71e25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>proximal_gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCTCAGTGCATAAGGCCAAT</td>\n",
       "      <td>chr10.786_second_two</td>\n",
       "      <td>22422921</td>\n",
       "      <td>22422940</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21422921</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204683', 'ENSG00000204682', 'ENSG00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCCTCCCAATAACCCTACAA</td>\n",
       "      <td>chr10.779_second_two</td>\n",
       "      <td>22306605</td>\n",
       "      <td>22306624</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21306605</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204683', 'ENSG00000204682', 'ENSG00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTGCTGAAGATAAATGCA</td>\n",
       "      <td>chr10.836_top_two</td>\n",
       "      <td>23144140</td>\n",
       "      <td>23144121</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24144121</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGACCGCCGTGAAGACCCCG</td>\n",
       "      <td>chr10.790_top_two</td>\n",
       "      <td>22518200</td>\n",
       "      <td>22518219</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21518200</td>\n",
       "      <td>23519666</td>\n",
       "      <td>['ENSG00000204682', 'ENSG00000180592', 'ENSG00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence         g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC     chr10.845_top_two  23345017  23345036    chr10   \n",
       "1  CCTCAGTGCATAAGGCCAAT  chr10.786_second_two  22422921  22422940    chr10   \n",
       "2  GCCTCCCAATAACCCTACAA  chr10.779_second_two  22306605  22306624    chr10   \n",
       "3  GTTTGCTGAAGATAAATGCA     chr10.836_top_two  23144140  23144121    chr10   \n",
       "4  AGACCGCCGTGAAGACCCCG     chr10.790_top_two  22518200  22518219    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "1  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "2  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "3  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "4  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right                                proximal_gene_names  \n",
       "0  21519685  24345036  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...  \n",
       "1  21422921  23519666  ['ENSG00000204683', 'ENSG00000204682', 'ENSG00...  \n",
       "2  21306605  23519666  ['ENSG00000204683', 'ENSG00000204682', 'ENSG00...  \n",
       "3  21519685  24144121  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...  \n",
       "4  21518200  23519666  ['ENSG00000204682', 'ENSG00000180592', 'ENSG00...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e2833b35-a8c8-46fb-89e6-ef751e174dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_pairs_genes = guide_pairs_genes.explode('proximal_gene_names')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "35e40652-c33f-4857-98d8-214fd91a5e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(83720570, 13)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "82944055-3626-4d40-a271-3cc8b368c95c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>proximal_gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>'ENSG00000204682'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>'ENSG00000180592'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>'ENSG00000078403'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>'ENSG00000136770'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>'ENSG00000223601'</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence      g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right proximal_gene_names  \n",
       "0  21519685  24345036   'ENSG00000204682'  \n",
       "0  21519685  24345036   'ENSG00000180592'  \n",
       "0  21519685  24345036   'ENSG00000078403'  \n",
       "0  21519685  24345036   'ENSG00000136770'  \n",
       "0  21519685  24345036   'ENSG00000223601'  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9e229633-401b-4c4a-a0b0-8f4a5de3c7cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g1_spacer_sequence</th>\n",
       "      <th>g1_gRNA_group</th>\n",
       "      <th>g1_start</th>\n",
       "      <th>g1_end</th>\n",
       "      <th>g1_chrom</th>\n",
       "      <th>g2_spacer_sequence</th>\n",
       "      <th>g2_gRNA_group</th>\n",
       "      <th>g2_start</th>\n",
       "      <th>g2_end</th>\n",
       "      <th>g2_chrom</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>proximal_gene_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>ENSG00000204682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>ENSG00000180592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>ENSG00000078403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>ENSG00000136770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAAAGCTCCTCCAGTTCAC</td>\n",
       "      <td>chr10.845_top_two</td>\n",
       "      <td>23345017</td>\n",
       "      <td>23345036</td>\n",
       "      <td>chr10</td>\n",
       "      <td>AAAGTTTGTTGAAAGTGGCG</td>\n",
       "      <td>chr10.791_top_two</td>\n",
       "      <td>22519685</td>\n",
       "      <td>22519666</td>\n",
       "      <td>chr10</td>\n",
       "      <td>21519685</td>\n",
       "      <td>24345036</td>\n",
       "      <td>ENSG00000223601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g1_spacer_sequence      g1_gRNA_group  g1_start    g1_end g1_chrom  \\\n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "0  AGAAAGCTCCTCCAGTTCAC  chr10.845_top_two  23345017  23345036    chr10   \n",
       "\n",
       "     g2_spacer_sequence      g2_gRNA_group  g2_start    g2_end g2_chrom  \\\n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "0  AAAGTTTGTTGAAAGTGGCG  chr10.791_top_two  22519685  22519666    chr10   \n",
       "\n",
       "       left     right proximal_gene_names  \n",
       "0  21519685  24345036     ENSG00000204682  \n",
       "0  21519685  24345036     ENSG00000180592  \n",
       "0  21519685  24345036     ENSG00000078403  \n",
       "0  21519685  24345036     ENSG00000136770  \n",
       "0  21519685  24345036     ENSG00000223601  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_pairs_genes['proximal_gene_names'] = guide_pairs_genes['proximal_gene_names'].apply(lambda x: x.strip(\"'\"))\n",
    "guide_pairs_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "31c24387-31c0-41c0-b5cf-105b3c3580dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_pairs_genes.to_csv('./../data/guide_pairs_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee094a23-18f9-47b7-8f13-45fad5aab207",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    1\n",
       "B    2\n",
       "C    3\n",
       "D    4\n",
       "E    5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_series = pd.Series([1, 2, 3, 4, 5], index=['A', 'B', 'C', 'D', 'E'])\n",
    "test_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c7459cc6-308b-4f48-905d-aeacadceeaa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "C     6\n",
       "D     7\n",
       "E     8\n",
       "F     9\n",
       "G    10\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_series_2 = pd.Series([6, 7, 8, 9, 10], index=['C', 'D', 'E', 'F', 'G'])\n",
    "test_series_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "41efaa6c-ba19-4e79-9abd-2fe32f61fdc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0\n",
       "A  1\n",
       "B  2\n",
       "C  3\n",
       "D  4\n",
       "E  5"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_1 = pd.DataFrame(test_series)\n",
    "test_df_2 = pd.DataFrame(test_series_2)\n",
    "test_df_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98f94836-e540-47aa-b7db-d0a75b458395",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0\n",
       "C   6\n",
       "D   7\n",
       "E   8\n",
       "F   9\n",
       "G  10"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "982c0803-c6b1-48ea-81cd-02eda940c286",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>0_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0_x  0_y\n",
       "C    3    6\n",
       "D    4    7\n",
       "E    5    8"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_1.merge(test_df_2, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "45aa1daa-ddd5-42c1-abf9-26b676af1c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "C    3\n",
       "D    4\n",
       "E    5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_series[test_series.index.isin(test_series_2.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f3ff6561-9333-42af-8818-da04581daab8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>AAACCGCTCCCGAGCACGGG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205792</th>\n",
       "      <td>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205793</th>\n",
       "      <td>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205794</th>\n",
       "      <td>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205795</th>\n",
       "      <td>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205796</th>\n",
       "      <td>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205797 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    cell  AAACCGCTCCCGAGCACGGG\n",
       "0       AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                     0\n",
       "1       AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                     0\n",
       "2       AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                     0\n",
       "3       AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                     0\n",
       "4       AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                     0\n",
       "...                                  ...                   ...\n",
       "205792  TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                     0\n",
       "205793  TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                     0\n",
       "205794  TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                     0\n",
       "205795  TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                     0\n",
       "205796  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                     0\n",
       "\n",
       "[205797 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_vec = pd.read_csv('../data/test_input_vector.csv')\n",
    "output_vec = pd.read_csv('../data/test_output_vector.csv')\n",
    "input_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ba996864-1d94-41d7-b750-66724ec638cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AAACCGCTCCCGAGCACGGG'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_gene_pairs = pd.read_csv('../data/guide_gene_pairs.csv')\n",
    "guide_gene_pairs.iloc[0]['spacer_sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b1e691ee-0c6d-4ec0-a3ba-ae1305ba6a9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell</th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTATATGAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACAACGATGG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGATAATG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAAAGCTCCTCCAGTTCAC</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATCGCTTTGACTGTGACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAATAAAGAACAGAACACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTAAATTGAGACCTCAGGAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTTCCCCCCACCAATAACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  205797 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell                  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 1   \n",
       "TGATCGCTTTGACTGTGACA                                 1   \n",
       "ACAATAAAGAACAGAACACA                                 1   \n",
       "GTAAATTGAGACCTCAGGAG                                 1   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTATATGAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2  ...  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  ...   \n",
       "TGATCGCTTTGACTGTGACA                                 0  ...   \n",
       "ACAATAAAGAACAGAACACA                                 0  ...   \n",
       "GTAAATTGAGACCTCAGGAG                                 0  ...   \n",
       "TCTTCCCCCCACCAATAACA                                 0  ...   \n",
       "\n",
       "cell                  TTTGTCACAACGATGG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGATAATG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  \n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  \n",
       "TGATCGCTTTGACTGTGACA                                 0  \n",
       "ACAATAAAGAACAGAACACA                                 0  \n",
       "GTAAATTGAGACCTCAGGAG                                 0  \n",
       "TCTTCCCCCCACCAATAACA                                 0  \n",
       "\n",
       "[5 rows x 205797 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_guide_matrix = pd.read_hdf('../data/cell_guide_matrix.h5')\n",
    "cell_guide_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "114221de-4c2d-4085-8977-cbee75f1955d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    4\n",
       "B    5\n",
       "C    6\n",
       "D    7\n",
       "E    8\n",
       "dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_1['test'] = [3, 3, 3, 3, 3]\n",
    "test_df_1.apply(sum, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e7f043-7637-40bb-887a-114927abff40",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py39]",
   "language": "python",
   "name": "conda-env-py39-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
