{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "696db547-1aa9-45e2-8f51-7cda23a46df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data analysis packages\n",
    "import numpy as np\n",
    "import pandas as pd \n",
    "\n",
    "# sparse matrix\n",
    "from scipy.io import mmread\n",
    "\n",
    "# modeling\n",
    "import statsmodels.api as sm\n",
    "from patsy import dmatrices\n",
    "from scipy import stats\n",
    "from scipy.stats import nbinom, chi2\n",
    "from scipy.optimize import minimize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "045c5dfd-1cc8-4323-a452-47ede7210a3e",
   "metadata": {},
   "source": [
    "## Define Null and Alternative Hypothesis Log-Likelihood Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d89dfa1-46ba-49f4-95d4-847aba595dda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_vec = pd.read_csv('../data/test_input_vector.csv')\n",
    "output_vec = pd.read_csv('../data/test_output_vector.csv')\n",
    "output_vec = output_vec.rename({'Unnamed: 0': 'cell'}, axis=1)\n",
    "merged_df = input_vec.merge(output_vec, on='cell')\n",
    "merged_df.iloc[:, 1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b11d0289-926b-45c4-aa69-838ed09f3763",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in test input and output vectors\n",
    "input_vector = merged_df.iloc[:, 1].values\n",
    "output_vector = merged_df.iloc[:, 2].values\n",
    "\n",
    "# preview input vector\n",
    "input_vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "10090728-9287-4313-ae87-9f2307c0779d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 1, 0])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview output vector\n",
    "output_vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "41190fdd-020c-413b-96e9-0a48885435f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nbinom_ll(params, x, counts, s):\n",
    "    '''\n",
    "    this function calculates the log-likelihood for the alternative hypothesis\n",
    "    alternative hypothesis - beta1 != 0 (indicating effect from guide presence on gene expr)\n",
    "    params[0] = beta0 (intercept term)\n",
    "    params[1] = beta1 (guide effect size)\n",
    "    params[2] = sqrt(disperion)\n",
    "    x1 = indicator vector (vector indicating whether guide is present in cells or not)\n",
    "    counts = vector of counts for UMIs of a gene in each cell (representative of gene expression)\n",
    "    s = scaling factor\n",
    "    '''\n",
    "    \n",
    "    # calculate value for dispersion of negative binomial distribution\n",
    "    disp = params[2] ** 2\n",
    "    \n",
    "    # calculate mu for negative binomial distribution\n",
    "    mu = np.exp(params[0] + x * params[1] + np.log(s))\n",
    "    \n",
    "    # calculate probability value for negative binomial distribution (parameterization of distribution)\n",
    "    prob = disp / (disp + mu)\n",
    "    \n",
    "    # calculate log-likelihood vector for observed counts under parameterized negative binomial distribution\n",
    "    ll = nbinom.logpmf(counts, n=disp, p=prob)\n",
    "    \n",
    "    # take sum of log likelihood vector\n",
    "    return ll.sum()\n",
    "\n",
    "def null_ll(params, counts, s):\n",
    "    '''\n",
    "    this function calculates the log-likelihood for the null hypothesis\n",
    "    null hypothesis - beta1 == 0 (indicating no effect from guide presence on gene expr)\n",
    "    params[0] = beta0 (intercept term)\n",
    "    params[1] = sqrt(dispersion)\n",
    "    counts = vector of counts for UMIs of a gene in each cell (representative of gene expression)\n",
    "    s = scaling factor\n",
    "    '''\n",
    "    \n",
    "    # calculate value for dispersion of negative binomial distribution\n",
    "    disp = params[1] ** 2\n",
    "    \n",
    "    # calculate mu for negative binomial distribution \n",
    "    mu = np.exp(params[0] + np.log(s))\n",
    "    \n",
    "    # calculate probability value for negative binomial distribution (parameterization of distribution)\n",
    "    prob = disp / (disp + mu)\n",
    "    \n",
    "    # calculate log-likelihood vector for observed counts under parameterized negative binomial distribution \n",
    "    ll = nbinom.logpmf(counts, n=disp, p=prob)\n",
    "    \n",
    "    # take sum of log-likelihood vector\n",
    "    return ll.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ce1a4109-65c8-42d3-8373-0bd10c6ffbf7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-181401.81470817997\n",
      "-181027.16970529922\n"
     ]
    }
   ],
   "source": [
    "# test negative binomial functions\n",
    "params = [0, 1, 1]\n",
    "print(nbinom_ll(params, input_vector, output_vector, 1))\n",
    "print(null_ll(params, output_vector, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3185222c-60c9-4238-a7a5-d9ac160734fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-749.2900057614897\n"
     ]
    }
   ],
   "source": [
    "# calculate likelihood ratio (test statistic)\n",
    "alt_ll = nbinom_ll(params, input_vector, output_vector, 1)\n",
    "base_ll = null_ll(params, output_vector, 1)\n",
    "ts = -2 * (base_ll - alt_ll)\n",
    "print(ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "56659154-0fb7-43ab-a425-c5ee74191015",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate p-value of likelihood ratio under chi-squared distribution\n",
    "pval = 1 - chi2.cdf(ts, 1)\n",
    "pval"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3820a3a5-2b27-4f89-9b97-329e64af18c7",
   "metadata": {},
   "source": [
    "## Calculate Scaling Factors (takes time to run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3d69a600-65b2-4b36-aa91-d5fc62e00933",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13135, 207324)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate scaling factors\n",
    "\n",
    "# read in count data\n",
    "count_matrix = mmread('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.exprs.mtx')\n",
    "count_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4a06c29-4fb0-4c1c-aa40-a55d1cf797a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>207314</th>\n",
       "      <th>207315</th>\n",
       "      <th>207316</th>\n",
       "      <th>207317</th>\n",
       "      <th>207318</th>\n",
       "      <th>207319</th>\n",
       "      <th>207320</th>\n",
       "      <th>207321</th>\n",
       "      <th>207322</th>\n",
       "      <th>207323</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 207324 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   0       1       2       3       4       5       6       7       8       \\\n",
       "0       0       0       0       0       0       0       0       0       0   \n",
       "1       1       0       0       0       0       0       0       0       0   \n",
       "2       1       0       1       1       1       0       0       0       0   \n",
       "3       0       0       0       1       0       0       0       0       0   \n",
       "4       0       0       1       0       0       0       0       1       0   \n",
       "\n",
       "   9       ...  207314  207315  207316  207317  207318  207319  207320  \\\n",
       "0       0  ...       0       0       0       0       0       0       0   \n",
       "1       0  ...       0       0       0       1       0       0       0   \n",
       "2       1  ...       0       0       0       0       1       0       0   \n",
       "3       0  ...       0       0       0       0       0       0       0   \n",
       "4       0  ...       0       0       0       0       0       0       0   \n",
       "\n",
       "   207321  207322  207323  \n",
       "0       0       0       0  \n",
       "1       1       0       0  \n",
       "2       2       0       2  \n",
       "3       0       0       0  \n",
       "4       0       0       0  \n",
       "\n",
       "[5 rows x 207324 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_matrix_df = pd.DataFrame(count_matrix.toarray())\n",
    "count_matrix_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "346c55a5-4e9b-46a1-877f-b756db97ee78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2\n",
       "1         AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2\n",
       "2         AAACCTGCAAACAACA-1_1A_1_SI-GA-E2\n",
       "3         AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2\n",
       "4         AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2\n",
       "                        ...               \n",
       "207319    TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9\n",
       "207320    TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9\n",
       "207321    TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9\n",
       "207322    TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9\n",
       "207323    TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9\n",
       "Length: 207324, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in column names from corresponding cells file\n",
    "colnames = []\n",
    "with open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.cells.txt') as f:\n",
    "    colnames = f.readlines()\n",
    "\n",
    "colnames = pd.Series(colnames).str.strip()\n",
    "colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30cf5c09-27ec-417d-88f7-2e112cee22e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        ENSG00000238009\n",
       "1        ENSG00000237683\n",
       "2        ENSG00000228463\n",
       "3        ENSG00000237094\n",
       "4        ENSG00000235373\n",
       "              ...       \n",
       "13130    ENSG00000215689\n",
       "13131    ENSG00000215781\n",
       "13132    ENSG00000220023\n",
       "13133    ENSG00000215615\n",
       "13134    ENSG00000215699\n",
       "Length: 13135, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in index (row names) from corresponding genes file \n",
    "rownames = []\n",
    "with open('/iblm/netapp/data1/external/Gasperini2019/suppl/GSE120861_at_scale_screen.genes.txt') as f:\n",
    "    rownames = f.readlines()\n",
    "    \n",
    "rownames = pd.Series(rownames).str.strip()\n",
    "rownames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a59eef94-41f6-43bd-98ed-65d945883835",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTATATGAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACAACGATGG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGATAATG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237683</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237094</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000235373</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 207324 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 1   \n",
       "\n",
       "                 AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 1   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGGTATATGAG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 1   \n",
       "\n",
       "                 AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2  ...  \\\n",
       "ENSG00000238009                                 0  ...   \n",
       "ENSG00000237683                                 0  ...   \n",
       "ENSG00000228463                                 1  ...   \n",
       "ENSG00000237094                                 0  ...   \n",
       "ENSG00000235373                                 0  ...   \n",
       "\n",
       "                 TTTGTCACAACGATGG-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCACAGATAATG-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 1   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 1   \n",
       "ENSG00000228463                                 2   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  \\\n",
       "ENSG00000238009                                 0   \n",
       "ENSG00000237683                                 0   \n",
       "ENSG00000228463                                 0   \n",
       "ENSG00000237094                                 0   \n",
       "ENSG00000235373                                 0   \n",
       "\n",
       "                 TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  \n",
       "ENSG00000238009                                 0  \n",
       "ENSG00000237683                                 0  \n",
       "ENSG00000228463                                 2  \n",
       "ENSG00000237094                                 0  \n",
       "ENSG00000235373                                 0  \n",
       "\n",
       "[5 rows x 207324 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_matrix_df.index = rownames\n",
    "count_matrix_df.columns = colnames\n",
    "count_matrix_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "35c7e811-3509-462f-aef9-d68bc9e0c17e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2    17566\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2     8917\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2    14626\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2    22783\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2    10124\n",
       "                                    ...  \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9    17934\n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9    16542\n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    14992\n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9     5149\n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9     9499\n",
       "Length: 207324, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate UMI counts per cell\n",
    "umi_counts_per_cell = count_matrix_df.sum()\n",
    "umi_counts_per_cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "db06b24e-b230-431a-81b3-0fa4d0a737a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3767321066"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate total UMI counts\n",
    "total_umi_count = umi_counts_per_cell.sum()\n",
    "total_umi_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "88694f03-047b-4d5f-9141-9f64a7909cc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([17566,  8917, 14626, ..., 14992,  5149,  9499])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umi_counts_per_cell.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "aad52844-3902-4537-babd-318ed77bf696",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4.66272975e-06, 2.36693392e-06, 3.88233435e-06, ...,\n",
       "       3.97948562e-06, 1.36675370e-06, 2.52142035e-06])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate UMI count proportions (scaling factors)\n",
    "scaling_factors = umi_counts_per_cell.values / total_umi_count\n",
    "scaling_factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d11c9b56-58fb-408a-81c8-c658a82736b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "207324"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(scaling_factors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a8d87a64-802b-4b15-a65d-864a578c4617",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2    0.000005\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2    0.000002\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2    0.000004\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2    0.000006\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2    0.000003\n",
       "                                      ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9    0.000005\n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9    0.000004\n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    0.000004\n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9    0.000001\n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9    0.000003\n",
       "Length: 207324, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert scaling factors into a series\n",
    "scaling_factor_series = pd.Series(data=scaling_factors, index=colnames)\n",
    "scaling_factor_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "298350ed-f1ce-4324-a4db-0f868293efd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write series to a CSV file\n",
    "scaling_factor_series.to_csv('../data/scaling_factors.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d92ec5b-133f-45c6-87d5-e382d7862aab",
   "metadata": {},
   "source": [
    "## Calculate Test Initial Parameter Estimates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ebd3d9eb-f86f-44aa-907f-6a8ea53667ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 1, 0])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview output vector\n",
    "output_vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "19c2ab0d-4896-4d54-8782-7752de1d9f6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaling_factor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  scaling_factor\n",
       "cell                                            \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2        0.000005\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2        0.000002\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2        0.000004\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2        0.000006\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2        0.000003"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in scaling factor values \n",
    "scaling_factors = pd.read_csv('../data/scaling_factors.csv')\n",
    "scaling_factors.columns = ['cell', 'scaling_factor']\n",
    "scaling_factors = scaling_factors.set_index('cell')\n",
    "scaling_factors.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "99904216-7b62-49ee-ac39-54c1320edd0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell</th>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTATATGAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACAACGATGG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGATAATG-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAAAGCTCCTCCAGTTCAC</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATCGCTTTGACTGTGACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAATAAAGAACAGAACACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTAAATTGAGACCTCAGGAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTTCCCCCCACCAATAACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 205797 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell                  AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 1   \n",
       "TGATCGCTTTGACTGTGACA                                 1   \n",
       "ACAATAAAGAACAGAACACA                                 1   \n",
       "GTAAATTGAGACCTCAGGAG                                 1   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGCGCAA-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTAGGGACT-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTATATGAG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  AAACCTGGTCAAAGCG-1_1A_1_SI-GA-E2  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 1   \n",
       "\n",
       "cell                  AAACCTGGTCTTCAAG-1_1A_1_SI-GA-E2  ...  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  ...   \n",
       "TGATCGCTTTGACTGTGACA                                 0  ...   \n",
       "ACAATAAAGAACAGAACACA                                 0  ...   \n",
       "GTAAATTGAGACCTCAGGAG                                 0  ...   \n",
       "TCTTCCCCCCACCAATAACA                                 0  ...   \n",
       "\n",
       "cell                  TTTGTCACAACGATGG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACACTTCTGC-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGATAATG-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACAGCCAGAA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCACATTAGGCT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  \\\n",
       "AGAAAGCTCCTCCAGTTCAC                                 0   \n",
       "TGATCGCTTTGACTGTGACA                                 0   \n",
       "ACAATAAAGAACAGAACACA                                 0   \n",
       "GTAAATTGAGACCTCAGGAG                                 0   \n",
       "TCTTCCCCCCACCAATAACA                                 0   \n",
       "\n",
       "cell                  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  \n",
       "AGAAAGCTCCTCCAGTTCAC                                 0  \n",
       "TGATCGCTTTGACTGTGACA                                 0  \n",
       "ACAATAAAGAACAGAACACA                                 0  \n",
       "GTAAATTGAGACCTCAGGAG                                 0  \n",
       "TCTTCCCCCCACCAATAACA                                 0  \n",
       "\n",
       "[5 rows x 205797 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in cell-guide matrix\n",
    "cell_guide_matrix = pd.read_hdf('./../data/cell_guide_matrix.h5')\n",
    "cell_guide_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ab53879e-cba2-465b-a738-76acb1951eca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell\n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2    0.000005\n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2    0.000002\n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2    0.000004\n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2    0.000006\n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2    0.000003\n",
       "                                      ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9    0.000005\n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9    0.000004\n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    0.000004\n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9    0.000001\n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9    0.000003\n",
       "Name: scaling_factor, Length: 205797, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_scaling_factors = scaling_factors.merge(cell_guide_matrix.iloc[0], on='cell')\n",
    "merged_scaling_factors = merged_scaling_factors.iloc[:, 0]\n",
    "merged_scaling_factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c69f6ab-5010-4c6f-a197-1b99f9f48ef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get numpy array of scaling factors (for faster computation)\n",
    "scaling_factors_arr = merged_scaling_factors.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a27e047a-1dbd-4b42-a48d-021f4f934050",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get subsets of output vector and scaling factors based on guide presence in cells\n",
    "cells_with_guide_for_gene = output_vector[input_vector.astype(bool)]\n",
    "cells_wo_guide_for_gene = output_vector[~input_vector.astype(bool)]\n",
    "scaling_factors_with_guide_for_gene = scaling_factors_arr[input_vector.astype(bool)]\n",
    "scaling_factors_wo_guide_for_gene = scaling_factors_arr[~input_vector.astype(bool)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ed8cd42d-81cc-4942-9bd5-7458a3565123",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate initial parameter estimates\n",
    "beta0_estimate = np.mean(np.log(cells_wo_guide_for_gene + 1) - np.log(scaling_factors_wo_guide_for_gene))\n",
    "beta1_estimate = beta0_estimate - np.mean(np.log(cells_with_guide_for_gene + 1) \n",
    "                                          - np.log(scaling_factors_with_guide_for_gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fc40a53-b9f9-4bfe-b183-b2c3c29858be",
   "metadata": {},
   "source": [
    "## Test Optimization (MLE) for Single Guide-Gene Pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "70d9e0ea-4cad-4d50-aef2-833c97934ae0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " final_simplex: (array([[ 2.32854235e+001, -7.45834073e-155],\n",
       "       [ 2.32854235e+001, -7.45834074e-155],\n",
       "       [ 2.32854235e+001, -7.45834074e-155]]), array([-31486173.72933589, -31486173.72926591, -31486173.7292487 ]))\n",
       "           fun: -31486173.729335893\n",
       "       message: 'Optimization terminated successfully.'\n",
       "          nfev: 1611\n",
       "           nit: 890\n",
       "        status: 0\n",
       "       success: True\n",
       "             x: array([ 2.32854235e+001, -7.45834073e-155])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test null model optimization\n",
    "null = minimize(null_ll, x0=(beta0_estimate, beta1_estimate), args=(output_vector, scaling_factors_arr),\n",
    "         method='Nelder-Mead', options={'maxiter': 1e5, 'maxfev':1e5})\n",
    "null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b3535632-0a35-42ed-a893-ab295b4575d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " final_simplex: (array([[ 2.30146991e+001, -8.64656093e-005, -7.45834073e-155],\n",
       "       [ 2.30147785e+001, -8.64199114e-005, -7.45834073e-155],\n",
       "       [ 2.30147285e+001, -8.64474675e-005, -7.45834073e-155],\n",
       "       [ 2.30147273e+001, -8.64508930e-005, -7.45834073e-155]]), array([-31486173.72935445, -31486173.72935445, -31486173.72935445,\n",
       "       -31486173.72935445]))\n",
       "           fun: -31486173.729354452\n",
       "       message: 'Optimization terminated successfully.'\n",
       "          nfev: 173\n",
       "           nit: 77\n",
       "        status: 0\n",
       "       success: True\n",
       "             x: array([ 2.30146991e+001, -8.64656093e-005, -7.45834073e-155])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test alternative model optimization\n",
    "alt = minimize(nbinom_ll, x0=(null.x[0], 0, null.x[1]), args=(input_vector, output_vector, scaling_factors_arr),\n",
    "               method='Nelder-Mead', options={'maxiter': 1e5, 'maxfev': 1e5})\n",
    "alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "83a699a1-bd48-4836-ba37-5f623eeb7afb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "205797"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(output_vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d3ea60d-13ac-42e4-b27b-08a5c1ca88f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1\n",
       "1    2\n",
       "2    3\n",
       "3    4\n",
       "4    5\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_series = pd.Series([1, 2, 3, 4, 5])\n",
    "test_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f20c22a7-b5a1-43af-9009-c3d944756922",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "::\n",
      "\n",
      "    Number of Observations - 303 (counties in California).\n",
      "\n",
      "    Number of Variables - 13 and 8 interaction terms.\n",
      "\n",
      "    Definition of variables names::\n",
      "\n",
      "        NABOVE   - Total number of students above the national median for the\n",
      "                   math section.\n",
      "        NBELOW   - Total number of students below the national median for the\n",
      "                   math section.\n",
      "        LOWINC   - Percentage of low income students\n",
      "        PERASIAN - Percentage of Asian student\n",
      "        PERBLACK - Percentage of black students\n",
      "        PERHISP  - Percentage of Hispanic students\n",
      "        PERMINTE - Percentage of minority teachers\n",
      "        AVYRSEXP - Sum of teachers' years in educational service divided by the\n",
      "                number of teachers.\n",
      "        AVSALK   - Total salary budget including benefits divided by the number\n",
      "                   of full-time teachers (in thousands)\n",
      "        PERSPENK - Per-pupil spending (in thousands)\n",
      "        PTRATIO  - Pupil-teacher ratio.\n",
      "        PCTAF    - Percentage of students taking UC/CSU prep courses\n",
      "        PCTCHRT  - Percentage of charter schools\n",
      "        PCTYRRND - Percentage of year-round schools\n",
      "\n",
      "        The below variables are interaction terms of the variables defined\n",
      "        above.\n",
      "\n",
      "        PERMINTE_AVYRSEXP\n",
      "        PEMINTE_AVSAL\n",
      "        AVYRSEXP_AVSAL\n",
      "        PERSPEN_PTRATIO\n",
      "        PERSPEN_PCTAF\n",
      "        PTRATIO_PCTAF\n",
      "        PERMINTE_AVTRSEXP_AVSAL\n",
      "        PERSPEN_PTRATIO_PCTAF\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(sm.datasets.star98.NOTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d6a0bee-c968-47fd-9822-a77f35385526",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = sm.datasets.star98.load(as_pandas=False)\n",
    "data.exog = sm.add_constant(data.exog, prepend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30a4fad6-f3e3-47c3-91df-554d019d0499",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(data.endog, data.exog, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f79371-c3e1-4d93-8dd3-d97e7170ee7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test using statsmodels\n",
    "test_series = pd.Series([1, 2, 3, 4, 5])\n",
    "glm_binom = sm.GLM(test_series, test_series, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5ce2b2c7-71bf-4a80-9302-d81d4d337d2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[23.28542348801676, -7.45834073276011e-155]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(list(null.x))\n",
    "list(null.x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b5585c45-0448-4cb9-aa6b-1a36310939ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9999991334588176"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# perform likelihood ratio test\n",
    "null_likelihood = -1 * null.fun\n",
    "alt_likelihood = -1 * alt.fun\n",
    "ts = -2 * (np.log(null_likelihood) - np.log(alt_likelihood))\n",
    "pval = 1 - chi2.cdf(ts, 1)\n",
    "pval"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0d2919d-a14f-4a44-a906-0d80102bbf96",
   "metadata": {},
   "source": [
    "## Scale Optimization to all Guide-Gene Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d2e16b21-31c5-4e42-8dd8-73ba7e64e908",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>genomic_coords</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>evalue</th>\n",
       "      <th>gRNAgroup.chr</th>\n",
       "      <th>gRNAgroup.start</th>\n",
       "      <th>gRNAgroup.stop</th>\n",
       "      <th>gRNAgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>[26606551, 26606569, 'chr1', 0.032]</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>[26606660, 26606641, 'chr1', 0.008]</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr1</td>\n",
       "      <td>26605667</td>\n",
       "      <td>26605668</td>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>[10530445, 10530427, 'chr11', 0.032]</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>[10530690, 10530708, 'chr11', 0.032]</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.032</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10530735</td>\n",
       "      <td>10530736</td>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>[124195027, 124195008, 'chr8', 0.008]</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>0.008</td>\n",
       "      <td>chr8</td>\n",
       "      <td>124191287</td>\n",
       "      <td>124191288</td>\n",
       "      <td>FAM83A_TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence                         genomic_coords  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG    [26606551, 26606569, 'chr1', 0.032]   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG    [26606660, 26606641, 'chr1', 0.008]   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   [10530445, 10530427, 'chr11', 0.032]   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   [10530690, 10530708, 'chr11', 0.032]   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  [124195027, 124195008, 'chr8', 0.008]   \n",
       "\n",
       "         start          end  chrom  evalue gRNAgroup.chr gRNAgroup.start  \\\n",
       "0   26606551.0   26606569.0   chr1   0.032          chr1        26605667   \n",
       "1   26606660.0   26606641.0   chr1   0.008          chr1        26605667   \n",
       "2   10530445.0   10530427.0  chr11   0.032         chr11        10530735   \n",
       "3   10530690.0   10530708.0  chr11   0.032         chr11        10530735   \n",
       "4  124195027.0  124195008.0   chr8   0.008          chr8       124191287   \n",
       "\n",
       "  gRNAgroup.stop     gRNAgroup  \n",
       "0       26605668  SH3BGRL3_TSS  \n",
       "1       26605668  SH3BGRL3_TSS  \n",
       "2       10530736  MTRNR2L8_TSS  \n",
       "3       10530736  MTRNR2L8_TSS  \n",
       "4      124191288    FAM83A_TSS  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in group sequence df\n",
    "group_sequence_coords = pd.read_csv('./../data/group_sequence_coords.csv')\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "eae8df50-57bb-46ca-90f9-d06d38a4c056",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr.targetgene</th>\n",
       "      <th>start.targetgene</th>\n",
       "      <th>stop.targetgene</th>\n",
       "      <th>ENSG.targetgene</th>\n",
       "      <th>targetgene_short_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28034777</td>\n",
       "      <td>28034778</td>\n",
       "      <td>ENSG00000150051</td>\n",
       "      <td>MKX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28287976</td>\n",
       "      <td>28287977</td>\n",
       "      <td>ENSG00000169126</td>\n",
       "      <td>ARMC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28571017</td>\n",
       "      <td>28571018</td>\n",
       "      <td>ENSG00000150054</td>\n",
       "      <td>MPP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28821422</td>\n",
       "      <td>28821423</td>\n",
       "      <td>ENSG00000095787</td>\n",
       "      <td>WAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>28966271</td>\n",
       "      <td>28966272</td>\n",
       "      <td>ENSG00000095739</td>\n",
       "      <td>BAMBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr.targetgene  start.targetgene  stop.targetgene  ENSG.targetgene  \\\n",
       "0          chr10          28034777         28034778  ENSG00000150051   \n",
       "1          chr10          28287976         28287977  ENSG00000169126   \n",
       "2          chr10          28571017         28571018  ENSG00000150054   \n",
       "3          chr10          28821422         28821423  ENSG00000095787   \n",
       "4          chr10          28966271         28966272  ENSG00000095739   \n",
       "\n",
       "  targetgene_short_name  \n",
       "0                   MKX  \n",
       "1                 ARMC4  \n",
       "2                  MPP7  \n",
       "3                   WAC  \n",
       "4                 BAMBI  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in genes df\n",
    "genes = pd.read_csv('./../data/genes_coords.csv')\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "fb2aec94-d92d-4a11-9ef8-3045f5c6ed3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_close_genes(guide_row):\n",
    "    '''find close genes given guide information'''\n",
    "    \n",
    "    # get guide information\n",
    "    guide_start = guide_row['start']\n",
    "    guide_end = guide_row['end']\n",
    "    guide_chrom = guide_row['chrom']\n",
    "    \n",
    "    # filter genes that match chromosome\n",
    "    filtered_genes = genes[genes['chr.targetgene'] == guide_chrom]\n",
    "    \n",
    "    # filter genes within range of 100kb\n",
    "    left = guide_start - 100000\n",
    "    right = guide_end + 100000\n",
    "    filtered_genes = filtered_genes[(filtered_genes['start.targetgene'] > left) & (filtered_genes['stop.targetgene'] < right)]\n",
    "    \n",
    "    return list(filtered_genes['ENSG.targetgene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "396d3133-2ecb-40f5-9c5c-0bd865aa19a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence        start          end  chrom\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG   26606551.0   26606569.0   chr1\n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG   26606660.0   26606641.0   chr1\n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   10530445.0   10530427.0  chr11\n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   10530690.0   10530708.0  chr11\n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  124195027.0  124195008.0   chr8"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter necessary columns\n",
    "group_sequence_coords = group_sequence_coords[['gRNA_group', 'spacer_sequence', 'start', 'end', 'chrom']]\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "e0592f2a-12aa-4d36-a2df-ab06960bc7ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[ENSG00000188782, ENSG00000130695, ENSG0000014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>CGCAGGCCGCTCATGCTGGG</td>\n",
       "      <td>26606660.0</td>\n",
       "      <td>26606641.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[ENSG00000188782, ENSG00000130695, ENSG0000014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAATAGTGGGAAGATTCGTG</td>\n",
       "      <td>10530445.0</td>\n",
       "      <td>10530427.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>[ENSG00000133805, ENSG00000255823, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTRNR2L8_TSS</td>\n",
       "      <td>AAGCTGTTCGGTAGTAAGGG</td>\n",
       "      <td>10530690.0</td>\n",
       "      <td>10530708.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>[ENSG00000133805, ENSG00000255823, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM83A_TSS</td>\n",
       "      <td>AACACACCACGGAGGAGTGG</td>\n",
       "      <td>124195027.0</td>\n",
       "      <td>124195008.0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>[ENSG00000147689, ENSG00000189376, ENSG0000016...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence        start          end  chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG   26606551.0   26606569.0   chr1   \n",
       "1  SH3BGRL3_TSS  CGCAGGCCGCTCATGCTGGG   26606660.0   26606641.0   chr1   \n",
       "2  MTRNR2L8_TSS  AAATAGTGGGAAGATTCGTG   10530445.0   10530427.0  chr11   \n",
       "3  MTRNR2L8_TSS  AAGCTGTTCGGTAGTAAGGG   10530690.0   10530708.0  chr11   \n",
       "4    FAM83A_TSS  AACACACCACGGAGGAGTGG  124195027.0  124195008.0   chr8   \n",
       "\n",
       "                                      proximal_genes  \n",
       "0  [ENSG00000188782, ENSG00000130695, ENSG0000014...  \n",
       "1  [ENSG00000188782, ENSG00000130695, ENSG0000014...  \n",
       "2  [ENSG00000133805, ENSG00000255823, ENSG0000011...  \n",
       "3  [ENSG00000133805, ENSG00000255823, ENSG0000011...  \n",
       "4  [ENSG00000147689, ENSG00000189376, ENSG0000016...  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get proximal genes for each guide sequence\n",
    "group_sequence_coords['proximal_genes'] = group_sequence_coords.apply(find_close_genes, axis=1)\n",
    "group_sequence_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "4951ba8b-b840-480a-8e61-10d81f325530",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000188782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000130695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000142669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000158062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence       start         end chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "1  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "2  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "3  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "4  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "\n",
       "    proximal_genes  \n",
       "0  ENSG00000188782  \n",
       "1  ENSG00000130695  \n",
       "2  ENSG00000142669  \n",
       "3  ENSG00000158062  \n",
       "4  ENSG00000169442  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explode dataframe to get guide-gene pairs\n",
    "guide_gene_pairs = group_sequence_coords.explode('proximal_genes', ignore_index=True)\n",
    "guide_gene_pairs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "efb15c90-d9d5-4277-94a1-03f0fd2a5788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34817, 6)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get shape of guide gene pairs dataframe\n",
    "guide_gene_pairs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "27b4ed8d-1684-436e-b194-c3a6f333fc28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write guide gene pairs to a CSV file\n",
    "guide_gene_pairs.to_csv('../data/guide_gene_pairs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1e5c7a04-7a64-41ba-83e0-a47aaecde752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>spacer_sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proximal_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000130695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000142669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000158062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SH3BGRL3_TSS</td>\n",
       "      <td>AAACCGCTCCCGAGCACGGG</td>\n",
       "      <td>26606551.0</td>\n",
       "      <td>26606569.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000176092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gRNA_group       spacer_sequence       start         end chrom  \\\n",
       "0  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "1  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "2  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "3  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "4  SH3BGRL3_TSS  AAACCGCTCCCGAGCACGGG  26606551.0  26606569.0  chr1   \n",
       "\n",
       "    proximal_genes  \n",
       "0  ENSG00000130695  \n",
       "1  ENSG00000142669  \n",
       "2  ENSG00000158062  \n",
       "3  ENSG00000169442  \n",
       "4  ENSG00000176092  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_gene_pairs = pd.read_csv('../data/guide_gene_pairs.csv')\n",
    "guide_gene_pairs = guide_gene_pairs[guide_gene_pairs['proximal_genes'].isin(count_matrix_df.index)].reset_index(drop=True)\n",
    "guide_gene_pairs.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py39]",
   "language": "python",
   "name": "conda-env-py39-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
